// @sanity/vision/lib/components/ResultTable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _calendarDate = _interopRequireDefault(require("../util/calendarDate"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ResultTable extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      expanded: []
    };
  }

  handleToggleExpandRow(id) {
    var expanded = this.state.expanded;
    var currentIndex = expanded.indexOf(id);

    if (currentIndex === -1) {
      expanded.push(id);
    } else {
      expanded.splice(currentIndex, 1);
    }
  }

  getExpandRowHandler(id) {
    return () => this.handleToggleExpandRow(id);
  }

  isExpanded(id) {
    return this.state.expanded.indexOf(id) !== -1;
  }

  render() {
    var docs = this.props.documents;

    if (!docs.length) {
      return _react.default.createElement("div", {
        className: "no-results"
      }, _react.default.createElement("p", null, "No results found for query:"), _react.default.createElement("code", null, this.props.query));
    }

    return _react.default.createElement("table", {
      className: "pure-table pure-table-striped vision_result-table"
    }, _react.default.createElement("thead", null, _react.default.createElement("tr", null, _react.default.createElement("th", null, "ID"), _react.default.createElement("th", null, "Type"), _react.default.createElement("th", {
      title: "All dates in UTC"
    }, "Last modified"), _react.default.createElement("th", {
      title: "All dates in UTC"
    }, "Created"))), _react.default.createElement("tbody", null, docs.map(doc => _react.default.createElement("tr", {
      key: doc._id,
      onClick: this.getExpandRowHandler(doc._id)
    }, _react.default.createElement("td", null, doc._id), _react.default.createElement("td", null, doc._type), _react.default.createElement("td", null, (0, _calendarDate.default)(doc._updatedAt)), _react.default.createElement("td", null, (0, _calendarDate.default)(doc._createdAt))))));
  }

}

ResultTable.propTypes = {
  query: _propTypes.default.string.isRequired,
  documents: _propTypes.default.arrayOf(_propTypes.default.shape({
    _id: _propTypes.default.string.isRequired,
    _type: _propTypes.default.string.isRequired,
    _updatedAt: _propTypes.default.string.isRequired,
    _createdAt: _propTypes.default.string.isRequired
  }))
};
var _default = ResultTable;
exports.default = _default;
