// @sanity/preview/lib/components/PreviewFields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PreviewFields;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PreviewSubscriber = _interopRequireDefault(require("./PreviewSubscriber"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function arrify(val) {
  if (Array.isArray(val)) {
    return val;
  }

  return typeof val === undefined ? [] : [val];
}

function PreviewFields(props) {
  return _react.default.createElement(_PreviewSubscriber.default, {
    value: props.document,
    type: props.type,
    fields: arrify(props.fields)
  }, (_ref) => {
    var snapshot = _ref.snapshot;
    return _react.default.createElement("span", null, snapshot && props.children(snapshot));
  });
}

PreviewFields.propTypes = {
  document: _propTypes.default.object,
  fields: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.arrayOf(_propTypes.default.string)]),
  type: _propTypes.default.object.isRequired,
  children: _propTypes.default.func
};
