// @sanity/form-builder/lib/sanity/utils/convertPath.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toGradient = toGradient;
exports.toFormBuilder = toFormBuilder;

var _mutator = require("@sanity/mutator");

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var IS_NUMERIC = /^\d+$/;

function unquote(str) {
  return str.replace(/^['"]/, '').replace(/['"]$/, '');
}

function splitAttr(segment) {
  var _segment$split = segment.split('=='),
      _segment$split2 = _slicedToArray(_segment$split, 2),
      attr = _segment$split2[0],
      key = _segment$split2[1];

  return {
    [attr]: unquote(key)
  };
}

function coerce(segment) {
  return IS_NUMERIC.test(segment) ? Number(segment) : segment;
}

function parseGradientPath(focusPathStr) {
  return focusPathStr.split(/[[.\]]/g).filter(Boolean).map(seg => seg.includes('==') ? splitAttr(seg) : coerce(seg));
}

function toGradient(formBuilderPath) {
  return (0, _mutator.arrayToJSONMatchPath)(formBuilderPath);
}

function toFormBuilder(gradientPath) {
  return parseGradientPath(gradientPath);
}
