// @sanity/form-builder/lib/inputs/BlockEditor/DeleteButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DeleteButton extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClick", event => {
      this.props.onClick(event);
    });
  }

  render() {
    return _react.default.createElement(_default.default, {
      kind: "simple",
      color: "danger",
      icon: _trashIcon.default,
      onClick: this.handleClick
    }, this.props.children);
  }

}

exports.default = DeleteButton;
