// @sanity/components/lib/toggles/Checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _checkboxStyle = _interopRequireDefault(require("part:@sanity/components/toggles/checkbox-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Checkbox extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      hasFocus: false
    });

    _defineProperty(this, "handleFocus", event => {
      this.setState({
        hasFocus: true
      });

      if (this.props.onFocus) {
        this.props.onFocus(event);
      }
    });

    _defineProperty(this, "handleBlur", event => {
      this.setState({
        hasFocus: false
      });

      if (this.props.onBlur) {
        this.props.onBlur(event);
      }
    });

    _defineProperty(this, "setInput", el => {
      this._input = el;
    });
  }

  componentDidMount() {
    if (typeof value === 'undefined' && this._input) {
      this._input.indeterminate = true;
    }
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    var _this$props = this.props,
        disabled = _this$props.disabled,
        markers = _this$props.markers,
        checked = _this$props.checked,
        label = _this$props.label,
        description = _this$props.description,
        children = _this$props.children,
        readOnly = _this$props.readOnly,
        rest = _objectWithoutProperties(_this$props, ["disabled", "markers", "checked", "label", "description", "children", "readOnly"]);

    var hasFocus = this.state.hasFocus;
    var checkedClass = checked ? _checkboxStyle.default.isChecked : _checkboxStyle.default.unChecked;

    if (typeof checked === 'undefined') {
      checkedClass = _checkboxStyle.default.indeterminate;
    }

    return _react.default.createElement("label", {
      title: label,
      className: "\n          ".concat(disabled || readOnly ? _checkboxStyle.default.isDisabled : _checkboxStyle.default.isEnabled, "\n          ").concat(checkedClass, "\n          ").concat(disabled || readOnly ? _checkboxStyle.default.isDisabled : _checkboxStyle.default.isEnabled, "\n          ").concat(checked ? _checkboxStyle.default.isChecked : _checkboxStyle.default.unChecked, "\n          ").concat(hasFocus ? _checkboxStyle.default.hasFocus : '', "\n        "),
      onBlur: this.handleBlur
    }, _react.default.createElement("input", _extends({}, rest, {
      className: _checkboxStyle.default.input,
      type: "checkbox",
      disabled: disabled || readOnly,
      checked: checked,
      ref: this.setInput,
      onFocus: this.handleFocus
    })), _react.default.createElement("div", {
      className: _checkboxStyle.default.label
    }, children || label), description && _react.default.createElement("div", {
      className: _checkboxStyle.default.description
    }, description), _react.default.createElement("div", {
      className: _checkboxStyle.default.focusHelper
    }), _react.default.createElement("div", {
      className: _checkboxStyle.default.boxOutline
    }, _react.default.createElement("div", {
      className: _checkboxStyle.default.tickOutline
    })), _react.default.createElement("div", {
      className: _checkboxStyle.default.tickHelper
    }), _react.default.createElement("div", {
      className: _checkboxStyle.default.rippleContainer
    }));
  }

}

exports.default = Checkbox;

_defineProperty(Checkbox, "propTypes", {
  label: _propTypes.default.string,
  description: _propTypes.default.string,
  markers: _propTypes.default.array,
  checked: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  readOnly: _propTypes.default.bool,
  children: _propTypes.default.any,
  onFocus: _propTypes.default.func,
  onBlur: _propTypes.default.func
});
