// @sanity/components/lib/toggles/ToggleButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _buttonStyle = _interopRequireDefault(require("part:@sanity/components/toggles/button-style"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ToggleButton extends _react.default.Component {
  render() {
    var _this$props = this.props,
        disabled = _this$props.disabled,
        selected = _this$props.selected,
        icon = _this$props.icon,
        className = _this$props.className,
        title = _this$props.title;
    var buttonClasses = "\n      ".concat(selected ? _buttonStyle.default.selected : _buttonStyle.default.unSelected, "\n      ").concat(className, "\n    ");
    return _react.default.createElement(_default.default, {
      className: buttonClasses,
      icon: icon,
      title: title,
      disabled: disabled,
      onClick: this.props.onClick,
      kind: "simple"
    }, this.props.children);
  }

}

exports.default = ToggleButton;

_defineProperty(ToggleButton, "propTypes", {
  icon: _propTypes.default.func,
  onClick: _propTypes.default.func.isRequired,
  selected: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  title: _propTypes.default.string
});

_defineProperty(ToggleButton, "defaultProps", {
  className: '',
  disabled: false
});
