// @sanity/default-layout/lib/components/ActionButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ActionModal = _interopRequireDefault(require("./ActionModal"));

var _fab = _interopRequireDefault(require("part:@sanity/components/buttons/fab"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ActionButton extends _react.default.Component {
  constructor(props) {
    super(props);
    this.handleToggleModal = this.handleToggleModal.bind(this);
    this.state = {
      modalOpen: false
    };
  }

  handleToggleModal() {
    this.setState({
      modalOpen: !this.state.modalOpen
    });
  }

  render() {
    if (this.state.modalOpen) {
      return _react.default.createElement(_ActionModal.default, {
        actions: this.props.actions,
        onClose: this.handleToggleModal
      });
    }

    return _react.default.createElement(_fab.default, {
      onClick: this.handleToggleModal,
      ripple: true,
      colored: true
    });
  }

}

_defineProperty(ActionButton, "propTypes", {
  actions: _propTypes.default.array
});

var _default = ActionButton;
exports.default = _default;
