// @sanity/default-layout/lib/components/CurrentVersionsDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./styles/UpdateNotifierDialog.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CurrentVersionsDialog(props) {
  var onClose = props.onClose,
      versions = props.versions;
  return _react.default.createElement(_default3.default, {
    isOpen: true,
    onClose: onClose,
    onClickOutside: onClose
  }, _react.default.createElement("div", {
    className: _UpdateNotifierDialog.default.content
  }, _react.default.createElement("div", null, _react.default.createElement("h2", null, "Studio is up to date"), _react.default.createElement("table", {
    className: _UpdateNotifierDialog.default.versionsTable
  }, _react.default.createElement("thead", null, _react.default.createElement("tr", null, _react.default.createElement("th", null, "Module"), _react.default.createElement("th", null, "Installed"), _react.default.createElement("th", null, "Latest"))), _react.default.createElement("tbody", null, Object.keys(versions).map(pkgName => _react.default.createElement("tr", {
    key: pkgName
  }, _react.default.createElement("td", null, pkgName), _react.default.createElement("td", null, versions[pkgName]), _react.default.createElement("td", null, versions[pkgName]))))), _react.default.createElement(_default2.default, {
    color: "primary",
    onClick: onClose
  }, "Close"))));
}

CurrentVersionsDialog.defaultProps = {
  versions: {}
};
CurrentVersionsDialog.propTypes = {
  onClose: _propTypes.default.func.isRequired,
  versions: _propTypes.default.objectOf(_propTypes.default.string)
};
var _default = CurrentVersionsDialog;
exports.default = _default;
