// @sanity/default-layout/lib/components/LoginStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactClickOutside = _interopRequireDefault(require("react-click-outside"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/menus/default"));

var _signOutIcon = _interopRequireDefault(require("part:@sanity/base/sign-out-icon"));

var _LoginStatus = _interopRequireDefault(require("./styles/LoginStatus.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class LoginStatus extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleClickOutside", () => {
      if (this.state.userMenuOpened) {
        this.setState({
          userMenuOpened: false
        });
      }
    });

    _defineProperty(this, "handleUserMenuToggle", () => {
      this.setState(state => {
        return {
          userMenuOpened: !state.userMenuOpened
        };
      });
    });

    _defineProperty(this, "handleUserMenuItemClick", item => {
      var onLogout = this.props.onLogout;

      if (item.action === 'signOut') {
        onLogout();
      }
    });

    this.state = {
      userMenuOpened: false
    };
  }

  render() {
    var user = this.props.user;
    var userMenuOpened = this.state.userMenuOpened;

    if (!user) {
      return null;
    }

    var className = _LoginStatus.default.root;
    if (this.props.className) className += this.props.className;
    return _react.default.createElement("div", {
      className: className
    }, _react.default.createElement("button", {
      className: _LoginStatus.default.button,
      onClick: this.handleUserMenuToggle,
      title: "Show user menu",
      type: "button"
    }, _react.default.createElement("div", {
      className: _LoginStatus.default.inner,
      tabIndex: -1
    }, user.profileImage ? _react.default.createElement("img", {
      src: user.profileImage,
      className: _LoginStatus.default.userImage,
      alt: "".concat(user.name, "'s profile image")
    }) : _react.default.createElement("div", {
      className: _LoginStatus.default.userImageMissing
    }, user.name ? user.name.charAt(0) : user.email.charAt(0)))), _react.default.createElement("div", {
      className: _LoginStatus.default.userName
    }, user.name), userMenuOpened && _react.default.createElement("div", {
      className: _LoginStatus.default.userMenu
    }, _react.default.createElement(_default2.default, {
      onAction: this.handleUserMenuItemClick,
      items: [{
        title: "Log out ".concat(user.name),
        icon: _signOutIcon.default,
        action: 'signOut'
      }],
      origin: "top-right",
      onClickOutside: this.handleUserMenuClose
    })));
  }

}

_defineProperty(LoginStatus, "propTypes", {
  className: _propTypes.default.string,
  onLogout: _propTypes.default.func,
  user: _propTypes.default.shape({
    name: _propTypes.default.string,
    profileImage: _propTypes.default.string
  })
});

var _default = (0, _reactClickOutside.default)(LoginStatus);

exports.default = _default;
