// @sanity/default-layout/lib/components/NavBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactTippy = require("react-tippy");

var _configSanity = _interopRequireDefault(require("config:sanity"));

var _plusIcon = _interopRequireDefault(require("part:@sanity/base/plus-icon"));

var _hamburgerIcon = _interopRequireDefault(require("part:@sanity/base/hamburger-icon"));

var _toolSwitcher = _interopRequireDefault(require("part:@sanity/default-layout/tool-switcher"));

var _searchIcon = _interopRequireDefault(require("part:@sanity/base/search-icon"));

var _router = require("part:@sanity/base/router");

var _spaces = require("../util/spaces");

var _Branding = _interopRequireDefault(require("./Branding"));

var _LoginStatus = _interopRequireDefault(require("./LoginStatus"));

var _SanityStatusContainer = _interopRequireDefault(require("./SanityStatusContainer"));

var _SearchContainer = _interopRequireDefault(require("./SearchContainer"));

var _SpaceSwitcher = _interopRequireDefault(require("./SpaceSwitcher"));

var _NavBar = _interopRequireDefault(require("./styles/NavBar.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function NavBar(props) {
  var searchIsOpen = props.searchIsOpen,
      onCreateButtonClick = props.onCreateButtonClick,
      onToggleMenu = props.onToggleMenu,
      onSwitchTool = props.onSwitchTool,
      onUserLogout = props.onUserLogout,
      onSearchOpen = props.onSearchOpen,
      onSearchClose = props.onSearchClose,
      onSetLoginStatusElement = props.onSetLoginStatusElement,
      onSetSearchElement = props.onSetSearchElement,
      router = props.router,
      tools = props.tools,
      user = props.user,
      showLabel = props.showLabel,
      showToolSwitcher = props.showToolSwitcher;
  var searchClassName = _NavBar.default.search;
  if (searchIsOpen) searchClassName += " ".concat(_NavBar.default.searchIsOpen);
  var className = _NavBar.default.root;
  if (showToolSwitcher) className += " ".concat(_NavBar.default.withToolSwitcher);
  return _react.default.createElement("div", {
    className: className,
    "data-search-open": searchIsOpen
  }, _react.default.createElement("div", {
    className: _NavBar.default.hamburger
  }, _react.default.createElement("button", {
    className: _NavBar.default.hamburgerButton,
    type: "button",
    title: "Open menu",
    onClick: onToggleMenu
  }, _react.default.createElement("div", {
    className: _NavBar.default.hamburgerButtonInner,
    tabIndex: -1
  }, _react.default.createElement(_hamburgerIcon.default, null)))), _react.default.createElement(_router.StateLink, {
    toIndex: true,
    className: _NavBar.default.branding
  }, _react.default.createElement(_Branding.default, {
    projectName: _configSanity.default && _configSanity.default.project.name
  })), _react.default.createElement("button", {
    className: _NavBar.default.createButton,
    onClick: onCreateButtonClick,
    type: "button"
  }, _react.default.createElement(_reactTippy.Tooltip, {
    disabled: 'ontouchstart' in document.documentElement,
    title: "Create new document",
    arrow: true,
    inertia: true,
    theme: "dark",
    distance: "7",
    sticky: true,
    size: "small"
  }, _react.default.createElement("div", {
    className: _NavBar.default.createButtonInner,
    tabIndex: -1
  }, _react.default.createElement("div", {
    className: _NavBar.default.createButtonIcon
  }, _react.default.createElement(_plusIcon.default, null)), _react.default.createElement("span", {
    className: _NavBar.default.createButtonText
  }, "New")))), _react.default.createElement("div", {
    className: searchClassName,
    ref: onSetSearchElement
  }, _react.default.createElement("div", null, _react.default.createElement(_SearchContainer.default, {
    shouldBeFocused: searchIsOpen,
    onOpen: onSearchOpen,
    onClose: onSearchClose
  }))), _spaces.HAS_SPACES && _react.default.createElement("div", {
    className: _NavBar.default.spaceSwitcher
  }, _react.default.createElement(_SpaceSwitcher.default, {
    isVisible: showToolSwitcher
  })), _react.default.createElement("div", {
    className: _NavBar.default.toolSwitcher
  }, _react.default.createElement(_toolSwitcher.default, {
    direction: "horizontal",
    isVisible: showToolSwitcher,
    tools: tools,
    activeToolName: router.state.tool,
    onSwitchTool: onSwitchTool,
    showLabel: showLabel
  })), _react.default.createElement("div", {
    className: _NavBar.default.extras
  }), _react.default.createElement("div", {
    className: _NavBar.default.sanityStatus
  }, _react.default.createElement(_SanityStatusContainer.default, null)), _react.default.createElement("div", {
    className: _NavBar.default.loginStatus,
    ref: onSetLoginStatusElement
  }, _react.default.createElement(_LoginStatus.default, {
    onLogout: onUserLogout,
    user: user
  })), _react.default.createElement("button", {
    className: _NavBar.default.searchButton,
    onClick: onSearchOpen,
    type: "button"
  }, _react.default.createElement("div", {
    className: _NavBar.default.searchButtonInner,
    tabIndex: -1
  }, _react.default.createElement("span", {
    className: _NavBar.default.searchButtonIcon
  }, _react.default.createElement(_searchIcon.default, null)), _react.default.createElement("span", {
    className: _NavBar.default.searchButtonText
  }, "Search"))));
}

NavBar.defaultProps = {
  showLabel: true,
  showToolSwitcher: true,
  onSetLoginStatusElement: undefined,
  onSetSearchElement: undefined
};
NavBar.propTypes = {
  searchIsOpen: _propTypes.default.bool.isRequired,
  onCreateButtonClick: _propTypes.default.func.isRequired,
  onToggleMenu: _propTypes.default.func.isRequired,
  onSwitchTool: _propTypes.default.func.isRequired,
  onUserLogout: _propTypes.default.func.isRequired,
  onSearchOpen: _propTypes.default.func.isRequired,
  onSearchClose: _propTypes.default.func.isRequired,
  onSetLoginStatusElement: _propTypes.default.func,
  onSetSearchElement: _propTypes.default.func,
  router: _propTypes.default.shape({
    state: _propTypes.default.shape({
      tool: _propTypes.default.string
    }),
    navigate: _propTypes.default.func
  }).isRequired,
  tools: _propTypes.default.arrayOf(_propTypes.default.shape({
    name: _propTypes.default.string
  })).isRequired,
  user: _propTypes.default.shape({
    name: _propTypes.default.string,
    profileImage: _propTypes.default.string
  }).isRequired,
  showLabel: _propTypes.default.bool,
  showToolSwitcher: _propTypes.default.bool
};
var _default = NavBar;
exports.default = _default;
