// @sanity/default-layout/lib/components/RenderTool.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _tool = _interopRequireDefault(require("all:part:@sanity/base/tool"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _RenderTool = _interopRequireDefault(require("./styles/RenderTool.css"));

var _ErrorBorkImage = _interopRequireDefault(require("./ErrorBorkImage"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getErrorWithStack(err) {
  var stack = err.stack.toString();
  var message = err.message;
  return stack.indexOf(message) === -1 ? "".concat(message, "\n\n").concat(stack) : stack;
}

function limitStackLength(stack) {
  return stack.split('\n').slice(0, 15).join('\n');
}

function formatStack(stack) {
  return stack // Prettify builder functions
  .replace(/\(\.\.\.\)\./g, '(...)\n  .') // Remove webpack cruft from function names
  .replace(/__WEBPACK_IMPORTED_MODULE_\d+_+/g, '') // Remove default export postfix from function names
  .replace(/___default\./g, '.') // Replace full host path, leave only path to JS-file
  .replace(new RegExp(" \\(https?:\\/\\/".concat(window.location.host), 'g'), ' (');
}

class RenderTool extends _react.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      error: null,
      showErrorDetails: __DEV__
    });

    _defineProperty(this, "handleShowDetails", () => {
      this.setState({
        showErrorDetails: true
      });
    });

    _defineProperty(this, "handleRetry", () => {
      this.setState({
        error: null
      });
    });
  }

  componentDidCatch(error, info) {
    this.setState({
      error: {
        error,
        info
      }
    });
  }

  renderError() {
    var tool = this.getActiveTool() || {};
    var _this$state$error = this.state.error,
        error = _this$state$error.error,
        info = _this$state$error.info;
    var showErrorDetails = this.state.showErrorDetails;
    return _react.default.createElement("div", {
      className: _RenderTool.default.error
    }, _react.default.createElement("div", {
      className: _RenderTool.default.errorSplash
    }, _react.default.createElement(_ErrorBorkImage.default, null), _react.default.createElement("h2", {
      className: _RenderTool.default.errorHeading
    }, "Rats! The ", tool.title || tool.name || 'active', " tool crashed!"), _react.default.createElement("div", {
      className: _RenderTool.default.errorActionButton
    }, _react.default.createElement(_default.default, {
      onClick: this.handleRetry
    }, "Retry"), _react.default.createElement(_default.default, {
      onClick: this.handleShowDetails,
      disabled: showErrorDetails
    }, "Details"))), showErrorDetails && _react.default.createElement("div", {
      className: _RenderTool.default.errorDetails
    }, _react.default.createElement("div", {
      className: _RenderTool.default.errorStackTraceWrapper
    }, _react.default.createElement("h3", null, "Stack trace:"), _react.default.createElement("pre", {
      className: _RenderTool.default.errorStackTrace
    }, formatStack(limitStackLength(getErrorWithStack(error))))), _react.default.createElement("div", {
      className: _RenderTool.default.errorComponentStack
    }, _react.default.createElement("h3", null, "Component stack:"), _react.default.createElement("pre", null, info.componentStack.replace(/^\s*\n+/, '')))));
  }

  getActiveTool() {
    var activeToolName = this.props.tool;

    var activeTool = _tool.default.find(tool => tool.name === activeToolName);

    return activeTool;
  }

  render() {
    if (this.state.error) {
      return this.renderError();
    }

    if (!_tool.default.length) {
      return _react.default.createElement("div", null, "No tools fulfills the part ", _react.default.createElement("code", null, "`part:@sanity/base/tool`"));
    }

    var activeTool = this.getActiveTool();

    if (!activeTool) {
      return _react.default.createElement("div", null, "Tool not found: ", this.props.tool);
    }

    var ActiveTool = activeTool.component;
    return _react.default.createElement(ActiveTool, this.props);
  }

}

exports.default = RenderTool;

_defineProperty(RenderTool, "propTypes", {
  tool: _propTypes.default.string
});

_defineProperty(RenderTool, "defaultProps", {
  tool: null
});
