// @sanity/default-layout/lib/components/SanityStatusContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _user = _interopRequireDefault(require("part:@sanity/base/user"));

var _reactPropsStream = require("react-props-stream");

var _versionChecker = _interopRequireDefault(require("part:@sanity/base/version-checker"));

var _sanityVersions = _interopRequireDefault(require("sanity:versions"));

var _SanityStatus = _interopRequireDefault(require("./SanityStatus"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// eslint-disable-next-line no-console
var levels = ['low', 'medium', 'high'];

var getHighestLevel = outdated => outdated.reduce((acc, pkg) => Math.max(acc, levels.indexOf(pkg.severity)), 0);

class SanityStatusContainer extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      showDialog: false
    });

    _defineProperty(this, "handleHideDialog", () => {
      this.setState({
        showDialog: false
      });
    });

    _defineProperty(this, "handleShowDialog", () => {
      this.setState({
        showDialog: true
      });
    });
  }

  render() {
    if (!this.props.showStatus) {
      return null;
    }

    var outdated = this.props.versionReply.outdated;
    var level = levels[getHighestLevel(outdated || [])];
    return _react.default.createElement(_SanityStatus.default, _extends({}, this.props.versionReply, {
      level: level,
      showDialog: this.state.showDialog,
      onHideDialog: this.handleHideDialog,
      onShowDialog: this.handleShowDialog,
      versions: _sanityVersions.default
    }));
  }

}

var _default = (0, _reactPropsStream.withPropsStream)(_user.default.currentUser.pipe((0, _operators.flatMap)(event => event.user && event.user.role === 'administrator' ? _versionChecker.default.checkVersions().then((_ref) => {
  var result = _ref.result;
  return {
    versionReply: result,
    showStatus: true
  };
}) : {
  showStatus: false
}), (0, _operators.catchError)(err => (0, _rxjs.of)({
  error: err,
  showStatus: false
}))), SanityStatusContainer);

exports.default = _default;
