// @sanity/default-layout/lib/components/SanityStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityStatus;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _packageIcon = _interopRequireDefault(require("part:@sanity/base/package-icon"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog"));

var _CurrentVersionsDialog = _interopRequireDefault(require("./CurrentVersionsDialog"));

var _SanityStatus = _interopRequireDefault(require("./styles/SanityStatus.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function formatUpdateLabel(len) {
  if (len === 1) return ' 1 update';
  return "".concat(len, " updates");
}

function SanityStatus(props) {
  var isSupported = props.isSupported,
      isUpToDate = props.isUpToDate,
      level = props.level,
      onHideDialog = props.onHideDialog,
      onShowDialog = props.onShowDialog,
      outdated = props.outdated,
      showDialog = props.showDialog,
      versions = props.versions;
  var currentLevel = outdated.length ? level : 'notice';
  var severity = isSupported ? currentLevel : 'high';
  var className = "".concat(_SanityStatus.default.root, " ").concat(_SanityStatus.default[severity]);
  var Dialog = isUpToDate ? _CurrentVersionsDialog.default : _UpdateNotifierDialog.default;
  return _react.default.createElement("div", {
    className: className
  }, showDialog && _react.default.createElement(Dialog, {
    severity: severity,
    outdated: outdated,
    onClose: onHideDialog,
    versions: versions
  }), _react.default.createElement("button", {
    className: _SanityStatus.default.button,
    onClick: onShowDialog,
    type: "button"
  }, _react.default.createElement("div", {
    className: _SanityStatus.default.buttonInner,
    tabIndex: -1
  }, isUpToDate ? _react.default.createElement("span", null, "Up to date") : _react.default.createElement("span", null, _react.default.createElement(_packageIcon.default, null), " ", formatUpdateLabel(outdated.length)))));
}

SanityStatus.defaultProps = {
  outdated: []
};
SanityStatus.propTypes = {
  isUpToDate: _propTypes.default.bool.isRequired,
  isSupported: _propTypes.default.bool.isRequired,
  level: _propTypes.default.string.isRequired,
  onHideDialog: _propTypes.default.func.isRequired,
  onShowDialog: _propTypes.default.func.isRequired,
  outdated: _propTypes.default.array,
  showDialog: _propTypes.default.bool.isRequired,
  versions: _propTypes.default.any.isRequired
};
