// @sanity/default-layout/lib/components/ToolSwitcherWidget.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _ToolSwitcherWidget = _interopRequireDefault(require("./styles/ToolSwitcherWidget.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToolSwitcherWidget(props) {
  var tools = props.tools,
      direction = props.direction,
      activeToolName = props.activeToolName,
      renderItem = props.renderItem;
  var showIcon = true;
  var showLabel = direction === 'vertical';
  return _react.default.createElement("ul", {
    className: direction === 'horizontal' ? _ToolSwitcherWidget.default.rootHorizontal : _ToolSwitcherWidget.default.rootVertical
  }, tools.map(tool => {
    var itemClass = activeToolName === tool.name ? _ToolSwitcherWidget.default.activeItem : _ToolSwitcherWidget.default.item;
    return _react.default.createElement("li", {
      key: tool.name,
      className: itemClass
    }, renderItem(tool, showIcon, showLabel));
  }));
}

ToolSwitcherWidget.defaultProps = {
  tools: [],
  direction: 'horizontal'
};
ToolSwitcherWidget.propTypes = {
  activeToolName: _propTypes.default.string,
  renderItem: _propTypes.default.func,
  direction: _propTypes.default.oneOf(['vertical', 'horizontal']),
  tools: _propTypes.default.arrayOf(_propTypes.default.shape({
    name: _propTypes.default.string.isRequired,
    icon: _propTypes.default.func
  }))
};
var _default = ToolSwitcherWidget;
exports.default = _default;
