// @sanity/default-layout/lib/components/ToolSwitcherItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _pluginIcon = _interopRequireDefault(require("part:@sanity/base/plugin-icon"));

var _reactTippy = require("react-tippy");

var _ToolSwitcherItem = _interopRequireDefault(require("./styles/ToolSwitcherItem.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToolSwitcherItem(props) {
  var isActive = props.isActive,
      title = props.title,
      icon = props.icon,
      showIcon = props.showIcon,
      showLabel = props.showLabel,
      direction = props.direction;
  var Icon = icon || _pluginIcon.default;
  return _react.default.createElement(_reactTippy.Tooltip, {
    className: "".concat(isActive ? _ToolSwitcherItem.default.rootActive : _ToolSwitcherItem.default.root, " ").concat(direction === 'vertical' ? _ToolSwitcherItem.default.vertical : _ToolSwitcherItem.default.horizontal),
    title: title,
    arrow: true,
    inertia: true,
    theme: "dark",
    distance: "7",
    sticky: true,
    size: "small",
    disabled: showLabel || 'ontouchstart' in document.documentElement
  }, _react.default.createElement("div", {
    className: _ToolSwitcherItem.default.inner,
    tabIndex: -1
  }, showIcon && _react.default.createElement("div", {
    className: _ToolSwitcherItem.default.iconContainer
  }, _react.default.createElement(Icon, null)), showLabel && _react.default.createElement("div", {
    className: _ToolSwitcherItem.default.toolName
  }, title)));
}

ToolSwitcherItem.defaultProps = {
  isActive: false,
  showIcon: true,
  showLabel: true,
  direction: 'horizontal'
};
ToolSwitcherItem.propTypes = {
  direction: _propTypes.default.oneOf(['horizontal', 'vertical']),
  isActive: _propTypes.default.bool,
  title: _propTypes.default.string.isRequired,
  icon: _propTypes.default.func,
  showIcon: _propTypes.default.bool,
  showLabel: _propTypes.default.bool
};
var _default = ToolSwitcherItem;
exports.default = _default;
