// @sanity/default-layout/lib/components/UpdateNotifierDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./styles/UpdateNotifierDialog.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var upperFirst = str => "".concat(str.slice(0, 1).toUpperCase()).concat(str.slice(1));

class UpdateNotifierDialog extends _react.Component {
  renderTable() {
    var outdated = this.props.outdated;
    return _react.default.createElement("div", null, _react.default.createElement("table", {
      className: _UpdateNotifierDialog.default.versionsTable
    }, _react.default.createElement("thead", null, _react.default.createElement("tr", null, _react.default.createElement("th", null, "Module"), _react.default.createElement("th", null, "Installed"), _react.default.createElement("th", null, "Latest"), _react.default.createElement("th", null, "Importance"))), _react.default.createElement("tbody", null, outdated.map(pkg => _react.default.createElement("tr", {
      key: pkg.name
    }, _react.default.createElement("td", null, pkg.name), _react.default.createElement("td", null, pkg.version), _react.default.createElement("td", null, pkg.latest), _react.default.createElement("td", null, upperFirst(pkg.severity || 'low')))))), _react.default.createElement("p", {
      className: _UpdateNotifierDialog.default.upgradeText
    }, "To upgrade, run the Sanity CLI command ", _react.default.createElement("code", {
      className: _UpdateNotifierDialog.default.code
    }, "sanity upgrade"), ' ', "in your project folder."));
  }

  renderContactDeveloper() {
    var severity = this.props.severity;
    return _react.default.createElement("div", null, _react.default.createElement("p", null, "You are running an outdated studio."), severity === 'high' ? _react.default.createElement("p", null, "Please get in touch with your developers and ask them to upgrade it for you.") : _react.default.createElement("p", null, "Consider getting in touch with your developers and ask them to upgrade it for you."), _react.default.createElement("details", null, _react.default.createElement("summary", null, "Developer info"), this.renderTable()));
  }

  render() {
    var _this$props = this.props,
        severity = _this$props.severity,
        onClose = _this$props.onClose;
    return _react.default.createElement(_default3.default, {
      isOpen: true,
      onClose: onClose,
      onClickOutside: onClose
    }, _react.default.createElement("div", {
      className: _UpdateNotifierDialog.default.content
    }, _react.default.createElement("div", null, _react.default.createElement("h2", null, severity === 'low' ? 'New versions available' : 'Studio is outdated'), __DEV__ ? this.renderTable() : this.renderContactDeveloper(), _react.default.createElement(_default2.default, {
      color: "primary",
      onClick: onClose
    }, "Close"))));
  }

}

_defineProperty(UpdateNotifierDialog, "propTypes", {
  onClose: _propTypes.default.func.isRequired,
  severity: _propTypes.default.string.isRequired,
  outdated: _propTypes.default.arrayOf(_propTypes.default.shape({
    name: _propTypes.default.string,
    latest: _propTypes.default.string,
    severity: _propTypes.default.string
  }))
});

_defineProperty(UpdateNotifierDialog, "defaultProps", {
  outdated: []
});

var _default = UpdateNotifierDialog;
exports.default = _default;
