// @sanity/base/lib/util/draftUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isDraft = isDraft;
exports.isDraftId = isDraftId;
exports.isPublishedId = isPublishedId;
exports.getDraftId = getDraftId;
exports.getPublishedId = getPublishedId;
exports.createDraftFrom = createDraftFrom;
exports.newDraftFrom = newDraftFrom;
exports.createPublishedFrom = createPublishedFrom;
exports.removeDupes = removeDupes;
exports.DRAFTS_FOLDER = void 0;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DRAFTS_FOLDER = 'drafts';
exports.DRAFTS_FOLDER = DRAFTS_FOLDER;
var DRAFTS_PREFIX = "".concat(DRAFTS_FOLDER, ".");

function isDraft(document) {
  return isDraftId(document._id);
}

function isDraftId(id) {
  return id.startsWith(DRAFTS_PREFIX);
}

function isPublishedId(id) {
  return !isDraftId(id);
}

function getDraftId(id) {
  return isDraftId(id) ? id : DRAFTS_PREFIX + id;
}

function getPublishedId(id) {
  return isDraftId(id) ? id.slice(DRAFTS_PREFIX.length) : id;
}

function createDraftFrom(document) {
  return _objectSpread({
    _id: getDraftId(document._id)
  }, document);
}

function newDraftFrom(document) {
  return _objectSpread({
    _id: DRAFTS_PREFIX
  }, document);
}

function createPublishedFrom(document) {
  return _objectSpread({
    _id: getPublishedId(document._id)
  }, document);
} // Removes published documents that also has a draft


function removeDupes(documents) {
  var drafts = documents.map(doc => doc._id).filter(isDraftId);
  return documents.filter(doc => {
    var draftId = getDraftId(doc._id);
    var publishedId = getPublishedId(doc._id);
    var hasDraft = drafts.includes(draftId);
    var isPublished = doc._id === publishedId;
    return isPublished ? !hasDraft : true;
  });
}
