// @sanity/desk-tool/lib/components/ConfirmUnpublish.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _enhanceWithReferringDocuments = _interopRequireDefault(require("./enhanceWithReferringDocuments"));

var _DocTitle = _interopRequireDefault(require("./DocTitle"));

var _ReferringDocumentsList = _interopRequireDefault(require("./ReferringDocumentsList"));

var _class, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _default = (0, _enhanceWithReferringDocuments.default)((_temp = _class = class ConfirmUnpublish extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleAction", action => {
      var _this$props = this.props,
          onCancel = _this$props.onCancel,
          onConfirm = _this$props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }

      if (action.name === 'cancel') {
        onCancel();
      }
    });
  }

  render() {
    var _this$props2 = this.props,
        isCheckingReferringDocuments = _this$props2.isCheckingReferringDocuments,
        referringDocuments = _this$props2.referringDocuments,
        draft = _this$props2.draft,
        published = _this$props2.published,
        onCancel = _this$props2.onCancel;
    var hasReferringDocuments = referringDocuments.length > 0;
    var canContinue = !isCheckingReferringDocuments;
    var actions = [canContinue && {
      name: 'confirm',
      title: hasReferringDocuments ? 'Try to unpublish anyway' : 'Unpublish now'
    }, {
      name: 'cancel',
      title: 'Cancel',
      kind: 'secondary'
    }].filter(Boolean);
    var title = isCheckingReferringDocuments ? 'Checking…' : 'Confirm unpublish';
    return _react.default.createElement(_fullscreen.default, {
      isOpen: true,
      showHeader: true,
      centered: true,
      title: title,
      onClose: onCancel,
      onAction: this.handleAction,
      actions: actions
    }, isCheckingReferringDocuments && _react.default.createElement(_spinner.default, {
      message: "Looking for referring documents\u2026"
    }), hasReferringDocuments && _react.default.createElement("div", null, _react.default.createElement("h3", null, "Warning: Found", ' ', referringDocuments.length === 1 ? 'a document' : "".concat(referringDocuments.length, " documents"), ' ', "that refers to ", '"', _react.default.createElement(_DocTitle.default, {
      document: draft || published
    }), '"'), _react.default.createElement("p", null, "You may not be allowed to unpublish", ' "', _react.default.createElement(_DocTitle.default, {
      document: draft || published
    }), '" ', "as the following document", referringDocuments.length === 1 ? '' : 's', " refers to it:"), _react.default.createElement(_ReferringDocumentsList.default, {
      documents: referringDocuments
    })), !isCheckingReferringDocuments && !hasReferringDocuments && _react.default.createElement("div", null, _react.default.createElement("p", null, "Are you sure you would like to unpublish the document", ' ', _react.default.createElement("strong", null, _react.default.createElement(_DocTitle.default, {
      document: draft || published
    })), "?"), _react.default.createElement("h2", null, "Careful!"), _react.default.createElement("p", null, "If you unpublish, this document will no longer be available for the public, but it will not be deleted and can be published again later if you change your mind.")));
  }

}, _defineProperty(_class, "propTypes", {
  onCancel: _propTypes.default.func.isRequired,
  onConfirm: _propTypes.default.func.isRequired,
  published: _propTypes.default.object,
  draft: _propTypes.default.object,
  referringDocuments: _propTypes.default.array,
  isCheckingReferringDocuments: _propTypes.default.bool
}), _temp));

exports.default = _default;
