// @sanity/desk-tool/lib/components/ConfirmDiscard.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _DocTitle = _interopRequireDefault(require("./DocTitle"));

var _distance_in_words_to_now = _interopRequireDefault(require("date-fns/distance_in_words_to_now"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var CANCEL_ACTION = {
  name: 'cancel',
  title: 'Cancel',
  kind: 'secondary',
  secondary: true
};

class ConfirmDiscard extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleAction", action => {
      var _this$props = this.props,
          onCancel = _this$props.onCancel,
          onConfirm = _this$props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }

      if (action.name === 'cancel') {
        onCancel();
      }
    });
  }

  render() {
    var _this$props2 = this.props,
        draft = _this$props2.draft,
        published = _this$props2.published,
        onCancel = _this$props2.onCancel;
    var confirmAction = {
      name: 'confirm',
      title: "Discard ".concat(published ? 'changes' : 'draft'),
      color: 'danger'
    };
    return _react.default.createElement(_fullscreen.default, {
      isOpen: true,
      showHeader: true,
      title: "Discard ".concat(published ? 'changes' : 'document'),
      centered: true,
      onClose: onCancel,
      onAction: this.handleAction,
      actions: [confirmAction, CANCEL_ACTION]
    }, _react.default.createElement("p", null, "Are you sure you would like to discard ", published ? 'changes in' : 'the document', ' ', _react.default.createElement("strong", null, _react.default.createElement(_DocTitle.default, {
      document: draft || published
    })), "?"), _react.default.createElement("p", null, "This will", ' ', published ? "revert to the latest published version of this document (published ".concat((0, _distance_in_words_to_now.default)(published._updatedAt, {
      addSuffix: true
    }), ").") : 'delete it entirely and there is no going back.'));
  }

}

exports.default = ConfirmDiscard;

_defineProperty(ConfirmDiscard, "propTypes", {
  onCancel: _propTypes.default.func.isRequired,
  onConfirm: _propTypes.default.func.isRequired,
  published: _propTypes.default.object,
  draft: _propTypes.default.object
});
