// @sanity/desk-tool/lib/components/TimeAgo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _shallowEquals = _interopRequireDefault(require("shallow-equals"));

var _distance_in_words_to_now = _interopRequireDefault(require("date-fns/distance_in_words_to_now"));

var _format = _interopRequireDefault(require("date-fns/format"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class TimeAgo extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "update", () => {
      this.forceUpdate();
    });
  }

  componentDidMount() {
    this.start();
  }

  componentWillUnmount() {
    this.stop();
  }

  componentWillReceiveProps(nextProps) {
    if (!(0, _shallowEquals.default)(this.props, nextProps)) {
      this.stop();
      this.start();
    }
  }

  start() {
    this.update();
    this.intervalId = setInterval(this.update, this.props.refreshInterval);
  }

  stop() {
    clearInterval(this.intervalId);
  }

  render() {
    var timestamp = (0, _format.default)(this.props.time, 'MMM D, YYYY, h:mm A Z');
    return _react.default.createElement("span", {
      title: timestamp
    }, (0, _distance_in_words_to_now.default)(this.props.time, {
      addSuffix: true
    }));
  }

}

exports.default = TimeAgo;

_defineProperty(TimeAgo, "propTypes", {
  refreshInterval: _propTypes.default.number,
  time: _propTypes.default.string.isRequired
});

_defineProperty(TimeAgo, "defaultProps", {
  refreshInterval: 1000 * 30
});
