// @sanity/default-login/lib/ErrorDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ErrorDialog;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _content = _interopRequireDefault(require("part:@sanity/components/dialogs/content"));

var _CorsCheck = _interopRequireDefault(require("./CorsCheck"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ErrorDialog(props) {
  var isNetworkError = props.error.isNetworkError;
  return _react.default.createElement(_fullscreen.default, {
    color: "danger",
    title: "Error",
    isOpen: true,
    centered: true
  }, !isNetworkError && _react.default.createElement("p", null, props.error.message), isNetworkError && _react.default.createElement(_content.default, {
    size: "medium",
    padding: "none"
  }, _react.default.createElement("p", null, "An error occured while attempting to reach the Sanity API."), _react.default.createElement(_CorsCheck.default, null)), _react.default.createElement(_default.default, {
    onClick: props.onRetry
  }, "Retry"));
}

ErrorDialog.propTypes = {
  error: _propTypes.default.shape({
    isNetworkError: _propTypes.default.boolean,
    message: _propTypes.default.string.isRequired
  }).isRequired,
  onRetry: _propTypes.default.func.isRequired
};
