// @sanity/structure/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const List_1 = require("./List");
const documentTypeListItems_1 = require("./documentTypeListItems");
const MenuItem_1 = require("./MenuItem");
const ListItem_1 = require("./ListItem");
const MenuItemGroup_1 = require("./MenuItemGroup");
const DocumentList_1 = require("./DocumentList");
const Editor_1 = require("./Editor");
const SerializeError_1 = require("./SerializeError");
exports.SerializeError = SerializeError_1.SerializeError;
const Component_1 = require("./Component");
const DocumentListItem_1 = require("./DocumentListItem");
const StructureBuilder = {
    documentTypeList: documentTypeListItems_1.getDocumentTypeList,
    documentTypeListItem: documentTypeListItems_1.getDocumentTypeListItem,
    documentTypeListItems: documentTypeListItems_1.getDocumentTypeListItems,
    list: (spec) => new List_1.ListBuilder(spec),
    listItem: (spec) => new ListItem_1.ListItemBuilder(spec),
    menuItem: (spec) => new MenuItem_1.MenuItemBuilder(spec),
    menuItemGroup: (spec) => new MenuItemGroup_1.MenuItemGroupBuilder(spec),
    documentList: (spec) => new DocumentList_1.DocumentListBuilder(spec),
    documentListItem: (spec) => new DocumentListItem_1.DocumentListItemBuilder(spec),
    orderingMenuItem: (ordering) => MenuItem_1.getOrderingMenuItem(ordering),
    orderingMenuItemsForType: (type) => MenuItem_1.getOrderingMenuItemsForSchemaType(type),
    editor: (spec) => new Editor_1.EditorBuilder(spec),
    component: (spec) => {
        return typeof spec === 'function'
            ? new Component_1.ComponentBuilder().component(spec)
            : new Component_1.ComponentBuilder(spec);
    },
    divider: () => ({ id: lodash_1.uniqueId('__divider__'), type: 'divider' })
};
exports.StructureBuilder = StructureBuilder;
