// @sanity/structure/lib/MenuItemGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SerializeError_1 = require("./SerializeError");
function maybeSerializeMenuItemGroup(item, index, path) {
    return item instanceof MenuItemGroupBuilder ? item.serialize({ path, index }) : item;
}
exports.maybeSerializeMenuItemGroup = maybeSerializeMenuItemGroup;
class MenuItemGroupBuilder {
    constructor(spec) {
        this._id = spec ? spec.id : '';
        this._title = spec ? spec.title : '';
    }
    id(id) {
        return new MenuItemGroupBuilder({ id, title: this._title });
    }
    getId() {
        return this._id;
    }
    title(title) {
        return new MenuItemGroupBuilder({ id: this._id, title });
    }
    getTitle() {
        return this._title;
    }
    serialize(options = { path: [] }) {
        const { _id, _title } = this;
        if (!_id) {
            throw new SerializeError_1.SerializeError('`id` is required for a menu item group', options.path, options.index, _title).withHelpUrl(SerializeError_1.HELP_URL.ID_REQUIRED);
        }
        if (!_title) {
            throw new SerializeError_1.SerializeError('`title` is required for a menu item group', options.path, _id).withHelpUrl(SerializeError_1.HELP_URL.TITLE_REQUIRED);
        }
        return {
            id: _id,
            title: _title
        };
    }
}
exports.MenuItemGroupBuilder = MenuItemGroupBuilder;
