// @sanity/form-builder/lib/FormBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _FormBuilderInput = require("./FormBuilderInput");

var _FormBuilderContext = _interopRequireDefault(require("./FormBuilderContext"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Todo: consider deprecating this in favor of <FormBuilderContext ...><FormBuilderInput .../></FormBuilderContext>
class FormBuilder extends _react.default.Component {
  render() {
    var _this$props = this.props,
        schema = _this$props.schema,
        value = _this$props.value,
        type = _this$props.type,
        onChange = _this$props.onChange,
        resolveInputComponent = _this$props.resolveInputComponent,
        resolvePreviewComponent = _this$props.resolvePreviewComponent,
        patchChannel = _this$props.patchChannel;

    if (!schema) {
      throw new TypeError('You must provide a schema to <FormBuilder (...)');
    }

    return _react.default.createElement(_FormBuilderContext.default, {
      schema: schema,
      value: value,
      resolveInputComponent: resolveInputComponent,
      resolvePreviewComponent: resolvePreviewComponent,
      patchChannel: patchChannel
    }, _react.default.createElement(_FormBuilderInput.FormBuilderInput, {
      value: value,
      type: type,
      onChange: onChange,
      level: 0,
      isRoot: true
    }));
  }

}

exports.default = FormBuilder;

_defineProperty(FormBuilder, "createPatchChannel", _FormBuilderContext.default.createPatchChannel);

_defineProperty(FormBuilder, "propTypes", {
  value: _propTypes.default.any,
  schema: _propTypes.default.object.isRequired,
  type: _propTypes.default.object.isRequired,
  onChange: _propTypes.default.func.isRequired,
  patchChannel: _propTypes.default.shape({
    onPatch: _propTypes.default.func
  }).isRequired,
  resolveInputComponent: _propTypes.default.func.isRequired,
  resolvePreviewComponent: _propTypes.default.func.isRequired
});

_defineProperty(FormBuilder, "defaultProps", {
  value: undefined
});
