// @sanity/form-builder/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createFormBuilder = createFormBuilder;
Object.defineProperty(exports, "FormBuilder", {
  enumerable: true,
  get: function get() {
    return _FormBuilder.default;
  }
});
Object.defineProperty(exports, "FormBuilderContext", {
  enumerable: true,
  get: function get() {
    return _FormBuilderContext.default;
  }
});
Object.defineProperty(exports, "BlockEditor", {
  enumerable: true,
  get: function get() {
    return _BlockEditor.default;
  }
});
Object.defineProperty(exports, "ReferenceInput", {
  enumerable: true,
  get: function get() {
    return _ReferenceInput.default;
  }
});
Object.defineProperty(exports, "FileInput", {
  enumerable: true,
  get: function get() {
    return _FileInput.default;
  }
});
Object.defineProperty(exports, "ImageInput", {
  enumerable: true,
  get: function get() {
    return _ImageInput.default;
  }
});

var _FormBuilder = _interopRequireDefault(require("./FormBuilder"));

var _FormBuilderContext = _interopRequireDefault(require("./FormBuilderContext"));

var _BlockEditor = _interopRequireDefault(require("./inputs/BlockEditor"));

var _ReferenceInput = _interopRequireDefault(require("./inputs/ReferenceInput"));

var _FileInput = _interopRequireDefault(require("./inputs/FileInput"));

var _ImageInput = _interopRequireDefault(require("./inputs/ImageInput"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createFormBuilder() {
  throw new Error('The factory function createFormBuilder(...) has been removed. Please use <FormBuilder .../> instead');
}
