// @sanity/core/lib/actions/graphql/registerBabelLoader.js
"use strict";

var fs = require('fs');

var path = require('path');

module.exports = basePath => {
  var configPath = path.join(basePath, '.babelrc');
  var babelConfig;

  try {
    // eslint-disable-next-line no-sync
    babelConfig = JSON.parse(fs.readFileSync(configPath, 'utf8'));
  } catch (err) {
    if (err.code !== 'ENOENT') {
      /* eslint-disable no-console */
      console.warn('Failed to read babel config at %s:', configPath);
      console.warn(err);
      console.warn('\n\nFalling back to default babel config');
      /* eslint-enable no-console */
    }
  }

  if (!babelConfig) {
    try {
      // eslint-disable-next-line
      babelConfig = require(path.join(basePath, 'babel.config.js'));
    } catch (err) {
      // Ignore errors, use default configuration
      babelConfig = {
        compact: false,
        root: basePath,
        ignore: [path.join(basePath, 'node_modules')],
        test: /.*/,
        presets: ['@babel/preset-react', ['@babel/preset-env', {
          targets: {
            node: 'current'
          }
        }]],
        plugins: ['@babel/plugin-proposal-class-properties']
      };
    }
  }

  require('@babel/register')(babelConfig);
};
