// @sanity/components/lib/autocomplete/DefaultAutocomplete.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _defaultStyle = _interopRequireDefault(require("part:@sanity/components/autocomplete/default-style"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/lists/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DefaultAutocomplete extends _react.default.Component {
  constructor(props, context) {
    super(props, context);
    this.handleSelect = this.handleSelect.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleKeyPress() {// console.log('press')
  }

  handleSelect(item) {
    this.props.onSelect(item);
  }

  handleChange(event) {
    this.props.onChange(event);
  }

  componentWillMount() {
    this._inputId = this.props.id || (0, _uniqueId2.default)('Autocomplete');
  }

  render() {
    var _this$props = this.props,
        suggestions = _this$props.suggestions,
        label = _this$props.label,
        value = _this$props.value;
    var isOpen = suggestions.length > 1;
    return _react.default.createElement(_default.default, {
      className: "".concat(isOpen ? _defaultStyle.default.opened : _defaultStyle.default.closed),
      label: label,
      labelFor: this._inputId
    }, _react.default.createElement("div", {
      className: _defaultStyle.default.input
    }, _react.default.createElement(_default2.default, {
      id: this._inputId,
      value: value,
      onKeyPress: this.handleKeyPress,
      className: _defaultStyle.default.textField,
      onChange: this.handleChange
    })), _react.default.createElement("div", {
      className: _defaultStyle.default.suggestionsContainer
    }, _react.default.createElement(_default3.default, {
      items: suggestions,
      className: _defaultStyle.default.suggestions,
      onSelect: this.handleSelect
    })));
  }

}

exports.default = DefaultAutocomplete;

_defineProperty(DefaultAutocomplete, "propTypes", {
  label: _propTypes.default.string.isRequired,
  onChange: _propTypes.default.func,
  onSelect: _propTypes.default.func,
  value: _propTypes.default.string,
  hasError: _propTypes.default.bool,
  onKeyPress: _propTypes.default.func,
  placeholder: _propTypes.default.string,
  isClearable: _propTypes.default.bool,
  isOpen: _propTypes.default.bool,
  id: _propTypes.default.string,
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string
  })),
  suggestions: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string
  }))
});

_defineProperty(DefaultAutocomplete, "defaultProps", {
  value: '',

  onChange() {},

  onSelect() {},

  suggestions: []
});
