// @sanity/state-router/lib/components/withRouterHOC.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = withRouter;

var _react = _interopRequireDefault(require("react"));

var _internalRouterContextTypeCheck = _interopRequireDefault(require("./internalRouterContextTypeCheck"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const NO_CONTEXT_STATE = {
  state: {},
  navigate: state => {
    throw new Error(`Cannot navigate to the state ${JSON.stringify(state)}. No router found in context`);
  },
  navigateIntent: intentName => {
    throw new Error(`Cannot navigate to the intent ${intentName}. No router found in context`);
  }
};

function withRouter(Component) {
  var _class, _temp;

  return _temp = _class = class extends _react.default.Component {
    constructor(props, context) {
      super();

      _defineProperty(this, "unsubscribe", void 0);

      _defineProperty(this, "state", void 0);

      _defineProperty(this, "state", {
        routerState: {}
      });

      _defineProperty(this, "context", void 0);

      const __internalRouter = context.__internalRouter;

      if (__internalRouter) {
        this.state = {
          routerState: __internalRouter.getState()
        };
      }
    }

    componentWillMount() {
      const __internalRouter = this.context.__internalRouter;

      if (!__internalRouter) {
        return;
      }

      this.unsubscribe = __internalRouter.channel.subscribe(() => {
        this.setState({
          routerState: __internalRouter.getState()
        });
      });
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    render() {
      const internalRouter = this.context.__internalRouter;
      const router = internalRouter ? {
        state: this.state.routerState,
        navigate: internalRouter.navigate,
        navigateIntent: internalRouter.navigateIntent
      } : NO_CONTEXT_STATE;
      return _react.default.createElement(Component, _extends({}, this.props, {
        router: router
      }));
    }

  }, _defineProperty(_class, "displayName", `withRouter(${Component.displayName || Component.name})`), _defineProperty(_class, "contextTypes", {
    __internalRouter: _internalRouterContextTypeCheck.default
  }), _temp;
}
