// @sanity/components/lib/utilities/CaptureOutsideClicks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class CaptureOutsideClicks extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "hadMouseDown", false);

    _defineProperty(this, "handleMouseDown", event => {
      if (this._wrapperElement.contains(event.target)) {
        this.hadMouseDown = true;
      }
    });

    _defineProperty(this, "handleDocumentClick", event => {
      if (this.props.onClickOutside && this._wrapperElement && !this._wrapperElement.contains(event.target) && !this.hadMouseDown) {
        this.props.onClickOutside(event);
      }

      this.hadMouseDown = false;
    });

    _defineProperty(this, "setWrapperElement", element => {
      this._wrapperElement = element;
    });
  }

  componentWillMount() {
    document.addEventListener('mouseup', this.handleDocumentClick);
    document.addEventListener('mousedown', this.handleMouseDown);
  }

  componentWillUnmount() {
    document.removeEventListener('mouseup', this.handleDocumentClick);
    document.removeEventListener('mousedown', this.handleMouseDown);
  }

  render() {
    var _this$props = this.props,
        wrapperElement = _this$props.wrapperElement,
        onClickOutside = _this$props.onClickOutside,
        rest = _objectWithoutProperties(_this$props, ["wrapperElement", "onClickOutside"]);

    return _react.default.createElement(wrapperElement, _objectSpread({}, rest, {
      ref: this.setWrapperElement
    }));
  }

}

exports.default = CaptureOutsideClicks;

_defineProperty(CaptureOutsideClicks, "propTypes", {
  onClickOutside: _propTypes.default.func,
  wrapperElement: _propTypes.default.string
});

_defineProperty(CaptureOutsideClicks, "defaultProps", {
  wrapperElement: 'div'
});
