// @sanity/components/lib/utilities/Poppable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactPopper = require("react-popper");

var _Poppable = _interopRequireDefault(require("./styles/Poppable.css"));

var _Stacked = _interopRequireDefault(require("./Stacked"));

var _Escapable = _interopRequireDefault(require("./Escapable"));

var _CaptureOutsideClicks = _interopRequireDefault(require("./CaptureOutsideClicks"));

var _Portal = require("./Portal");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Poppable extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "popperNode", undefined);

    _defineProperty(this, "setPopperNode", node => {
      this.popperNode = node;
    });

    _defineProperty(this, "handleClickOutside", ev => {
      if (!this.popperNode || !ev.target) {
        return;
      }

      if (!this.popperNode.contains(ev.target)) {
        this.props.onClickOutside(ev);
      }
    });
  }

  render() {
    var _this$props = this.props,
        onEscape = _this$props.onEscape,
        onClickOutside = _this$props.onClickOutside,
        target = _this$props.target,
        children = _this$props.children,
        referenceClassName = _this$props.referenceClassName,
        popperClassName = _this$props.popperClassName,
        referenceElement = _this$props.referenceElement; // Undefined referenceElement causes Popper to think it is defined

    var popperPropHack = {};

    if (referenceElement) {
      popperPropHack.referenceElement = referenceElement;
    }

    return _react.default.createElement(_reactPopper.Manager, null, !referenceElement && _react.default.createElement(_reactPopper.Reference, null, (_ref) => {
      var ref = _ref.ref;
      return _react.default.createElement("div", {
        ref: ref,
        className: referenceClassName
      });
    }), children && _react.default.createElement(_Portal.Portal, null, _react.default.createElement(_Stacked.default, null, isActive => _react.default.createElement("div", {
      className: _Poppable.default.portal
    }, _react.default.createElement(_reactPopper.Popper, _extends({
      innerRef: this.setPopperNode,
      modifiers: this.props.modifiers,
      placement: this.props.placement,
      positionFixed: true
    }, popperPropHack), (_ref2) => {
      var ref = _ref2.ref,
          placement = _ref2.placement,
          style = _ref2.style;
      return _react.default.createElement("div", {
        ref: ref,
        style: style,
        "data-placement": placement,
        className: popperClassName
      }, _react.default.createElement(_react.Fragment, null, _react.default.createElement(_Escapable.default, {
        onEscape: isActive ? onEscape : undefined
      }), onClickOutside ? _react.default.createElement(_CaptureOutsideClicks.default, {
        onClickOutside: isActive ? this.handleClickOutside : undefined
      }, children) : children));
    })))));
  }

}

exports.default = Poppable;

_defineProperty(Poppable, "propTypes", {
  onEscape: _propTypes.default.func,
  onClickOutside: _propTypes.default.func,
  target: _propTypes.default.node,
  children: _propTypes.default.node,
  referenceClassName: _propTypes.default.string,
  // When requiring this file in node, Element is undefined (since it's a window global)
  referenceElement: typeof window === 'undefined' ? _propTypes.default.any : _propTypes.default.instanceOf(Element),
  placement: _propTypes.default.string,
  positionFixed: _propTypes.default.bool,
  popperClassname: _propTypes.default.string,
  modifiers: _propTypes.default.shape({
    preventOverflow: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    customStyle: _propTypes.default.object,
    flip: _propTypes.default.object,
    offset: _propTypes.default.object
  })
});

_defineProperty(Poppable, "defaultProps", {
  placement: 'bottom-start',
  modifiers: {
    preventOverflow: 'viewport'
  }
});
