// @sanity/form-builder/lib/sanity/inputs/client-adapters/mockUpload.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockUpload = mockUpload;

var _rxjs = require("rxjs");

function createProgressEvent(progress) {
  return {
    type: 'progress',
    stage: 'upload',
    percent: progress
  };
}

function simulateProgress() {
  var updateInterval = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 200;
  var speed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 2;
  return new _rxjs.Observable(observer => {
    var progress = 0;
    observer.next(progress);
    var interval = setInterval(next, updateInterval);
    return () => clearInterval(interval);

    function next() {
      progress = Math.min(100, progress + (speed + Math.random() * speed));
      observer.next(progress);

      if (progress === 100) {
        observer.complete();
      }
    }
  });
}

function mockUpload(assetDoc) {
  return simulateProgress(100 + Math.random() * 500, 10 + Math.random() * 50).map(createProgressEvent).mergeMap(event => {
    return _rxjs.Observable.of(...[event, event.percent === 100 && {
      type: 'complete',
      asset: assetDoc
    }].filter(Boolean));
  });
}
