// @sanity/core/lib/commands/cors/deleteCorsOriginCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var helpText = "\nExamples\n  sanity cors delete\n  sanity cors delete http://localhost:3000\n";
var _default = {
  name: 'delete',
  group: 'cors',
  signature: '[ORIGIN]',
  helpText,
  description: 'Delete an existing CORS-origin from your project',
  action: function () {
    var _action = _asyncToGenerator(function* (args, context) {
      var output = context.output,
          apiClient = context.apiClient;

      var _args$argsWithoutOpti = _slicedToArray(args.argsWithoutOptions, 1),
          origin = _args$argsWithoutOpti[0];

      var client = apiClient({
        requireUser: true,
        requireProject: true
      });
      var originId = yield promptForOrigin(origin, context);

      try {
        yield client.request({
          method: 'DELETE',
          uri: "/cors/".concat(originId)
        });
        output.print('Origin deleted');
      } catch (err) {
        throw new Error("Origin deletion failed:\n".concat(err.message));
      }
    });

    function action(_x, _x2) {
      return _action.apply(this, arguments);
    }

    return action;
  }()
};
exports.default = _default;

function promptForOrigin(_x3, _x4) {
  return _promptForOrigin.apply(this, arguments);
}

function _promptForOrigin() {
  _promptForOrigin = _asyncToGenerator(function* (specified, context) {
    var specifiedOrigin = specified && specified.toLowerCase();
    var prompt = context.prompt,
        apiClient = context.apiClient;
    var client = apiClient({
      requireUser: true,
      requireProject: true
    });
    var origins = yield client.request({
      url: '/cors'
    });

    if (specifiedOrigin) {
      var selected = origins.filter(origin => origin.origin.toLowerCase() === specifiedOrigin)[0];

      if (!selected) {
        throw new Error("Origin \"".concat(specified, " not found\""));
      }

      return selected.id;
    }

    var choices = origins.map(origin => ({
      value: origin.id,
      name: origin.origin
    }));
    return prompt.single({
      message: 'Select origin to delete',
      type: 'list',
      choices
    });
  });
  return _promptForOrigin.apply(this, arguments);
}
