// @sanity/util/lib/reduceConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _mergeWith2 = _interopRequireDefault(require("lodash/mergeWith"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sanityEnv = process.env.SANITY_ENV || 'production';
var apiHosts = {
  staging: 'https://api.sanity.work',
  development: 'http://api.sanity.wtf'
};
var processEnvConfig = {
  project: process.env.STUDIO_BASEPATH ? {
    basePath: process.env.STUDIO_BASEPATH
  } : {}
};

function merge(objValue, srcValue, key) {
  if (Array.isArray(objValue)) {
    return objValue.concat(srcValue);
  } // Pass on to default merging strategy


  return undefined;
}

var _default = function _default(rawConfig) {
  var env = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'development';
  var apiHost = apiHosts[sanityEnv];
  var sanityConf = apiHost ? {
    api: {
      apiHost
    }
  } : {};
  var envConfig = (rawConfig.env || {})[env] || {};
  var config = (0, _mergeWith2.default)({}, rawConfig, envConfig, sanityConf, processEnvConfig, merge);
  delete config.env;
  return config;
};

exports.default = _default;
