// @sanity/form-builder/lib/inputs/BooleanInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _switch = _interopRequireDefault(require("part:@sanity/components/toggles/switch"));

var _checkbox = _interopRequireDefault(require("part:@sanity/components/toggles/checkbox"));

var _PatchEvent = _interopRequireWildcard(require("../PatchEvent"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Todo: support indeterminate state, see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox
class BooleanInput extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_input", void 0);

    _defineProperty(this, "handleChange", event => {
      this.props.onChange(_PatchEvent.default.from((0, _PatchEvent.set)(event.currentTarget.checked)));
    });

    _defineProperty(this, "setInput", input => {
      this._input = input;
    });
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    var _this$props = this.props,
        value = _this$props.value,
        type = _this$props.type,
        readOnly = _this$props.readOnly,
        onFocus = _this$props.onFocus;
    var isCheckbox = type.options && type.options.layout === 'checkbox';
    return isCheckbox ? _react.default.createElement(_checkbox.default, {
      readOnly: readOnly,
      onChange: this.handleChange,
      checked: value,
      ref: this.setInput,
      description: type.description
    }, type.title) : _react.default.createElement(_switch.default, {
      readOnly: readOnly,
      checked: value,
      label: type.title,
      description: type.description,
      onChange: this.handleChange,
      onFocus: onFocus,
      ref: this.setInput
    });
  }

}

exports.default = BooleanInput;
