// @sanity/form-builder/lib/inputs/SelectInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/selects/default"));

var _radio = _interopRequireDefault(require("part:@sanity/components/selects/radio"));

var _PatchEvent = _interopRequireWildcard(require("../PatchEvent"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var EMPTY_ITEM = {
  title: '',
  value: undefined
};

function toSelectItems(list) {
  return typeof list[0] === 'string' ? list.map(item => ({
    title: item,
    value: item
  })) : list;
}

class StringSelect extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_input", void 0);

    _defineProperty(this, "name", (0, _uniqueId2.default)('RadioName'));

    _defineProperty(this, "handleChange", item => {
      var onChange = this.props.onChange;
      onChange(_PatchEvent.default.from((0, _PatchEvent.set)(typeof item === 'string' ? item : item.value)));
    });

    _defineProperty(this, "setInput", input => {
      this._input = input;
    });
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    var _this$props = this.props,
        value = _this$props.value,
        readOnly = _this$props.readOnly,
        markers = _this$props.markers,
        type = _this$props.type,
        level = _this$props.level,
        onFocus = _this$props.onFocus;
    var items = toSelectItems(type.options.list || []);
    var currentItem = items.find(item => item.value === value);
    var isRadio = type.options && type.options.layout === 'radio';
    return _react.default.createElement(_default2.default, {
      markers: markers,
      level: level,
      label: type.title,
      description: type.description
    }, isRadio ? // todo: make separate inputs
    _react.default.createElement(_radio.default, {
      name: this.name,
      legend: type.title,
      items: items,
      onChange: this.handleChange,
      onFocus: onFocus,
      value: currentItem,
      direction: type.options.direction || 'vertical',
      ref: this.setInput,
      readOnly: readOnly
    }) : _react.default.createElement(_default.default, {
      label: type.title,
      value: currentItem,
      placeholder: type.placeholder,
      onChange: this.handleChange,
      onFocus: onFocus,
      items: [EMPTY_ITEM].concat(items),
      ref: this.setInput,
      readOnly: readOnly
    }));
  }

}

exports.default = StringSelect;

_defineProperty(StringSelect, "defaultProps", {
  value: ''
});
