// @sanity/components/lib/validation/ValidationListItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ValidationListItem = _interopRequireDefault(require("./styles/ValidationListItem.css"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ValidationListItem extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleKeyPress", event => {
      if (event.key === 'Enter') {
        this.handleClick(event);
      }
    });

    _defineProperty(this, "handleClick", event => {
      var _this$props = this.props,
          marker = _this$props.marker,
          onClick = _this$props.onClick;

      if (onClick) {
        onClick(event, marker.path);
      }
    });

    _defineProperty(this, "setElement", element => {
      this._element = element;
    });
  }

  componentDidMount() {
    if (this.props.hasFocus) {
      this.focus();
    }
  }

  focus() {
    setTimeout(() => {
      this._element.focus();
    }, 200);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.hasFocus) {
      this.focus();
    }
  }

  render() {
    var _this$props2 = this.props,
        marker = _this$props2.marker,
        onClick = _this$props2.onClick,
        path = _this$props2.path,
        showLink = _this$props2.showLink,
        truncate = _this$props2.truncate;
    var shouldRenderLink = onClick && showLink;
    return _react.default.createElement("a", {
      ref: this.setElement,
      tabIndex: 0,
      onClick: this.handleClick,
      onKeyPress: this.handleKeyPress,
      className: "\n          ".concat(onClick ? _ValidationListItem.default.interactiveItem : _ValidationListItem.default.item, "\n          ").concat(_ValidationListItem.default[marker.level], "\n          ").concat(truncate ? _ValidationListItem.default.truncate : '', "\n        ")
    }, _react.default.createElement("span", {
      className: _ValidationListItem.default.icon
    }, _react.default.createElement(_warningIcon.default, null)), _react.default.createElement("div", {
      className: _ValidationListItem.default.content
    }, _react.default.createElement("span", {
      className: _ValidationListItem.default.path
    }, path), _react.default.createElement("span", {
      className: _ValidationListItem.default.message
    }, marker.item.message)), shouldRenderLink && _react.default.createElement("span", {
      className: _ValidationListItem.default.link,
      title: "View"
    }, _react.default.createElement(_linkIcon.default, null)));
  }

}

exports.default = ValidationListItem;

_defineProperty(ValidationListItem, "propTypes", {
  onClick: _propTypes.default.func,
  showLink: _propTypes.default.bool,
  path: _propTypes.default.string,
  hasFocus: _propTypes.default.bool,
  truncate: _propTypes.default.bool,
  marker: _propTypes.default.shape({
    path: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.shape({
      _key: _propTypes.default.string
    })])),
    type: _propTypes.default.string,
    level: _propTypes.default.string,
    item: _propTypes.default.any
  }).isRequired
});

_defineProperty(ValidationListItem, "defaultProps", {
  path: '',
  onClick: undefined,
  showLink: false
});
