// @sanity/desk-tool/lib/pane/Editor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _debounce2 = _interopRequireDefault(require("lodash/debounce"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _reactTippy = require("react-tippy");

var _router = require("part:@sanity/base/router");

var _preview = require("part:@sanity/base/preview");

var _draftUtils = require("part:@sanity/base/util/draft-utils");

var _documentActionUtils = require("part:@sanity/base/util/document-action-utils");

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _popover = _interopRequireDefault(require("part:@sanity/components/dialogs/popover"));

var _formBuilder = _interopRequireDefault(require("part:@sanity/form-builder"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _undoIcon = _interopRequireDefault(require("part:@sanity/base/undo-icon"));

var _publicIcon = _interopRequireDefault(require("part:@sanity/base/public-icon"));

var _visibilityOffIcon = _interopRequireDefault(require("part:@sanity/base/visibility-off-icon"));

var _binaryIcon = _interopRequireDefault(require("part:@sanity/base/binary-icon"));

var _contentCopyIcon = _interopRequireDefault(require("part:@sanity/base/content-copy-icon"));

var _document = _interopRequireDefault(require("part:@sanity/base/datastore/document"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/panes/default"));

var _afterEditorComponent = _interopRequireDefault(require("all:part:@sanity/desk-tool/after-editor-component"));

var _syncIcon = _interopRequireDefault(require("part:@sanity/base/sync-icon"));

var _checkIcon = _interopRequireDefault(require("part:@sanity/base/check-icon"));

var _circleCheckIcon = _interopRequireDefault(require("part:@sanity/base/circle-check-icon"));

var _default4 = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _resolveProductionUrl = _interopRequireDefault(require("part:@sanity/transitional/production-preview/resolve-production-url?"));

var _list = _interopRequireDefault(require("part:@sanity/components/validation/list"));

var _chevronDownIcon = _interopRequireDefault(require("part:@sanity/base/chevron-down-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _languageSelectComponent = _interopRequireDefault(require("part:@sanity/desk-tool/language-select-component?"));

var _filterFieldsFn = _interopRequireDefault(require("part:@sanity/desk-tool/filter-fields-fn?"));

var _hotkeys = _interopRequireDefault(require("part:@sanity/components/typography/hotkeys"));

var _copyDocument = _interopRequireDefault(require("../utils/copyDocument"));

var _ConfirmUnpublish = _interopRequireDefault(require("../components/ConfirmUnpublish"));

var _ConfirmDelete = _interopRequireDefault(require("../components/ConfirmDelete"));

var _InspectView = _interopRequireDefault(require("../components/InspectView"));

var _DocTitle = _interopRequireDefault(require("../components/DocTitle"));

var _TimeAgo = _interopRequireDefault(require("../components/TimeAgo"));

var _Editor = _interopRequireDefault(require("./styles/Editor.css"));

var _class, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function navigateUrl(url) {
  window.open(url);
}

var preventDefault = ev => ev.preventDefault(); // Want a nicer api for listen/unlisten


function listen(target, eventType, callback) {
  var useCapture = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  target.addEventListener(eventType, callback, useCapture);
  return function unlisten() {
    target.removeEventListener(eventType, callback, useCapture);
  };
}

var getDuplicateItem = (draft, published) => ({
  action: 'duplicate',
  title: 'Duplicate',
  icon: _contentCopyIcon.default,
  isDisabled: !draft && !published
});

var getDiscardItem = (draft, published, isLiveEditEnabled) => isLiveEditEnabled ? null : {
  action: 'discard',
  title: 'Discard changes…',
  icon: _undoIcon.default,
  isDisabled: !draft || !published
};

var getUnpublishItem = (draft, published, isLiveEditEnabled) => isLiveEditEnabled ? null : {
  action: 'unpublish',
  title: 'Unpublish…',
  icon: _visibilityOffIcon.default,
  isDisabled: !published
};

var getDeleteItem = (draft, published) => ({
  group: 'danger',
  action: 'delete',
  title: 'Delete…',
  icon: _trashIcon.default,
  danger: true,
  isDisabled: !draft && !published
});

var getInspectItem = (draft, published) => ({
  action: 'inspect',
  title: _react.default.createElement("span", {
    className: _Editor.default.menuItem
  }, "Inspect", ' ', _react.default.createElement("span", {
    className: _Editor.default.hotkey
  }, _react.default.createElement(_hotkeys.default, {
    keys: ['Ctrl', 'Alt', 'I']
  }))),
  icon: _binaryIcon.default,
  isDisabled: !(draft || published)
});

var getProductionPreviewItem = (draft, published) => {
  var snapshot = draft || published;

  if (!snapshot || !_resolveProductionUrl.default) {
    return null;
  }

  var previewUrl;

  try {
    previewUrl = (0, _resolveProductionUrl.default)(snapshot);
  } catch (error) {
    error.message = "An error was thrown while trying to get production preview url: ".concat(error.message); // eslint-disable-next-line no-console

    console.error(error);
    return null;
  }

  return previewUrl && {
    action: 'production-preview',
    title: _react.default.createElement("span", {
      className: _Editor.default.menuItem
    }, "Open preview", _react.default.createElement("span", {
      className: _Editor.default.hotkey
    }, _react.default.createElement(_hotkeys.default, {
      keys: ['Ctrl', 'Alt', 'O']
    }))),
    icon: _publicIcon.default,
    url: previewUrl
  };
};

var getMenuItems = (enabledActions, draft, published, isLiveEditEnabled) => [getProductionPreviewItem, enabledActions.includes('delete') && getDiscardItem, enabledActions.includes('delete') && getUnpublishItem, enabledActions.includes('create') && getDuplicateItem, getInspectItem, enabledActions.includes('delete') && getDeleteItem].filter(Boolean).map(fn => fn(draft, published, isLiveEditEnabled)).filter(Boolean);

var isValidationError = marker => marker.type === 'validation' && marker.level === 'error';

var INITIAL_STATE = {
  inspect: false,
  isMenuOpen: false,
  isCreatingDraft: false,
  showSavingStatus: false,
  showConfirmDiscard: false,
  showConfirmDelete: false,
  showConfirmUnpublish: false,
  showValidationTooltip: false,
  focusPath: [],
  filterField: () => true
};

var _default = (0, _router.withRouterHOC)(( // eslint-disable-next-line
_temp = _class = class Editor extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", INITIAL_STATE);

    _defineProperty(this, "setSavingStatus", (0, _debounce2.default)(() => {
      this.setState({
        showSavingStatus: false
      });
    }, 1500, {
      trailing: true
    }));

    _defineProperty(this, "handleFocus", path => {
      this.setState({
        focusPath: path
      });
    });

    _defineProperty(this, "handleBlur", () => {// do nothing
    });

    _defineProperty(this, "handleCreateCopy", () => {
      var _this$props = this.props,
          router = _this$props.router,
          draft = _this$props.draft,
          published = _this$props.published,
          paneIndex = _this$props.paneIndex;
      var prevId = (0, _draftUtils.getPublishedId)((draft || published)._id);
      var duplicatedDocument = this.isLiveEditEnabled() ? (0, _copyDocument.default)(published) : (0, _draftUtils.newDraftFrom)((0, _copyDocument.default)(draft || published));
      this.duplicate$ = _document.default.create(duplicatedDocument).subscribe(copied => {
        var copyDocId = (0, _draftUtils.getPublishedId)(copied._id);
        var newPanes = router.state.panes.map((prev, i) => i === paneIndex - 1 && prev === prevId ? copyDocId : prev);
        router.navigate(_objectSpread({}, router.state, {
          panes: newPanes
        }));
      });
    });

    _defineProperty(this, "handleEditAsActualType", () => {
      var _this$props2 = this.props,
          router = _this$props2.router,
          draft = _this$props2.draft,
          published = _this$props2.published;
      router.navigateIntent('edit', {
        id: (0, _draftUtils.getPublishedId)((draft || published)._id),
        type: draft._type || published._type
      });
    });

    _defineProperty(this, "handleChange", changeEvent => {
      var onChange = this.props.onChange;
      onChange(changeEvent);
    });

    _defineProperty(this, "handleRestore", () => {
      var deletedSnapshot = this.props.deletedSnapshot;
      this.props.onCreate(deletedSnapshot);
    });

    _defineProperty(this, "handleMenuToggle", evt => {
      evt.stopPropagation();
      this.setState(prevState => ({
        isMenuOpen: !prevState.isMenuOpen
      }));
    });

    _defineProperty(this, "handleMenuClose", evt => {
      this.setState({
        isMenuOpen: false
      });
    });

    _defineProperty(this, "handlePublishRequested", () => {
      var _this$props3 = this.props,
          markers = _this$props3.markers,
          validationPending = _this$props3.validationPending,
          onPublish = _this$props3.onPublish,
          draft = _this$props3.draft;

      if (!draft) {
        return;
      }

      var errors = markers.filter(isValidationError);
      var hasErrors = errors.length > 0;

      if (validationPending || hasErrors) {
        this.setState(prevState => ({
          showValidationTooltip: !prevState.showValidationTooltip
        }));
        return;
      }

      onPublish(draft);
    });

    _defineProperty(this, "handleCancelUnpublish", () => {
      this.setState({
        showConfirmUnpublish: false
      });
    });

    _defineProperty(this, "handleCancelDelete", () => {
      this.setState({
        showConfirmDelete: false
      });
    });

    _defineProperty(this, "handleCancelDiscard", () => {
      this.setState({
        showConfirmDiscard: false
      });
    });

    _defineProperty(this, "handleConfirmUnpublish", () => {
      var onUnpublish = this.props.onUnpublish;
      onUnpublish();
      this.setState({
        showConfirmUnpublish: false
      });
    });

    _defineProperty(this, "handleConfirmDiscard", () => {
      var _this$props4 = this.props,
          onDiscardDraft = _this$props4.onDiscardDraft,
          draft = _this$props4.draft;
      onDiscardDraft(draft);
      this.setState({
        showConfirmDiscard: false
      });
    });

    _defineProperty(this, "handleConfirmDelete", () => {
      var _this$props5 = this.props,
          onDelete = _this$props5.onDelete,
          onDiscardDraft = _this$props5.onDiscardDraft,
          published = _this$props5.published;

      if (published) {
        onDelete();
      } else {
        onDiscardDraft();
      }

      this.setState({
        showConfirmDelete: false
      });
    });

    _defineProperty(this, "handleHideInspector", () => {
      this.setState({
        inspect: false
      });
    });

    _defineProperty(this, "handleMenuAction", item => {
      if (item.action === 'production-preview') {
        navigateUrl(item.url);
      }

      if (item.action === 'delete') {
        this.setState({
          showConfirmDelete: true
        });
      }

      if (item.action === 'discard') {
        this.setState({
          showConfirmDiscard: true
        });
      }

      if (item.action === 'unpublish') {
        this.setState({
          showConfirmUnpublish: true
        });
      }

      if (item.action === 'duplicate') {
        this.handleCreateCopy();
      }

      if (item.action === 'inspect') {
        this.setState({
          inspect: true
        });
      }

      this.setState({
        isMenuOpen: false
      });
    });

    _defineProperty(this, "handleCloseValidationResults", () => {
      this.setState({
        showValidationTooltip: false
      });
    });

    _defineProperty(this, "handleToggleValidationResults", () => {
      this.setState(prevState => ({
        showValidationTooltip: !prevState.showValidationTooltip
      }));
    });

    _defineProperty(this, "renderPublishButtonTooltip", (errors, published) => {
      if (errors.length > 0) {
        return _react.default.createElement("span", null, "Fix errors before publishing");
      }

      return _react.default.createElement("span", {
        className: _Editor.default.menuItem
      }, published ? 'Publish changes' : 'Publish', errors.length < 1 && _react.default.createElement("span", {
        className: _Editor.default.hotkey
      }, _react.default.createElement(_hotkeys.default, {
        keys: ['Ctrl', 'Alt', 'P']
      })));
    });

    _defineProperty(this, "renderActions", (props, state) => {
      var _this$props6 = this.props,
          draft = _this$props6.draft,
          published = _this$props6.published,
          markers = _this$props6.markers,
          type = _this$props6.type,
          isReconnecting = _this$props6.isReconnecting;
      var _this$state = this.state,
          showSavingStatus = _this$state.showSavingStatus,
          showValidationTooltip = _this$state.showValidationTooltip;
      var value = draft || published;
      var validation = markers.filter(marker => marker.type === 'validation');
      var errors = validation.filter(marker => marker.level === 'error');
      var warnings = validation.filter(marker => marker.level === 'warning');
      return _react.default.createElement("div", {
        className: _Editor.default.paneFunctions
      }, _languageSelectComponent.default && _react.default.createElement(_languageSelectComponent.default, null), showSavingStatus && _react.default.createElement(_reactTippy.Tooltip, {
        className: _Editor.default.syncStatusSyncing,
        arrow: true,
        theme: "light",
        size: "small",
        distance: "0",
        title: "Syncing your content with the Sanity cloud"
      }, _react.default.createElement("span", {
        className: _Editor.default.syncSpinnerContainer
      }, _react.default.createElement("span", {
        className: _Editor.default.syncSpinner
      }, _react.default.createElement(_syncIcon.default, null)), "\xA0Syncing\u2026")), isReconnecting && _react.default.createElement(_reactTippy.Tooltip, {
        className: _Editor.default.syncStatusReconnecting,
        arrow: true,
        theme: "light",
        size: "small",
        distance: "0",
        title: "Connection lost. Reconnecting\u2026"
      }, _react.default.createElement("span", {
        className: _Editor.default.syncSpinnerContainer
      }, _react.default.createElement("span", {
        className: _Editor.default.syncSpinner
      }, _react.default.createElement(_syncIcon.default, null)), "\xA0Reconnecting\u2026")), value && !showSavingStatus && !isReconnecting && _react.default.createElement(_reactTippy.Tooltip, {
        className: _Editor.default.syncStatusSynced,
        arrow: true,
        theme: "light",
        size: "small",
        distance: "0",
        title: "Synced with the Sanity cloud"
      }, _react.default.createElement(_checkIcon.default, null), " Synced ", this.isLiveEditEnabled() && ' (live)'), (errors.length > 0 || warnings.length > 0) && _react.default.createElement(_reactTippy.Tooltip, {
        arrow: true,
        theme: "light noPadding",
        trigger: "click",
        position: "bottom",
        interactive: true,
        duration: 100,
        open: showValidationTooltip,
        onRequestClose: this.handleCloseValidationResults,
        html: _react.default.createElement(_list.default, {
          truncate: true,
          markers: validation,
          showLink: true,
          isOpen: showValidationTooltip,
          documentType: type,
          onClose: this.handleCloseValidationResults,
          onFocus: this.handleFocus
        })
      }, _react.default.createElement(_default2.default, {
        color: "danger",
        bleed: true,
        icon: _warningIcon.default,
        padding: "small",
        onClick: this.handleToggleValidationResults
      }, errors.length, _react.default.createElement("span", {
        style: {
          paddingLeft: '0.5em',
          display: 'flex'
        }
      }, _react.default.createElement(_chevronDownIcon.default, null)))));
    });

    _defineProperty(this, "renderPublishInfo", () => {
      var _this$props7 = this.props,
          draft = _this$props7.draft,
          markers = _this$props7.markers,
          published = _this$props7.published,
          isReconnecting = _this$props7.isReconnecting,
          isCreatingDraft = _this$props7.isCreatingDraft,
          isPublishing = _this$props7.isPublishing,
          isUnpublishing = _this$props7.isUnpublishing;
      var validation = markers.filter(marker => marker.type === 'validation');
      var errors = validation.filter(marker => marker.level === 'error');
      return _react.default.createElement(_react.default.Fragment, null, (isCreatingDraft || isPublishing || isUnpublishing) && _react.default.createElement("div", {
        className: _Editor.default.spinnerContainer
      }, isCreatingDraft && _react.default.createElement(_spinner.default, {
        center: true,
        message: "Making changes\u2026"
      }), isPublishing && _react.default.createElement(_spinner.default, {
        center: true,
        message: "Publishing\u2026"
      }), isUnpublishing && _react.default.createElement(_spinner.default, {
        center: true,
        message: "Unpublishing\u2026"
      })), _react.default.createElement("div", {
        className: draft && !this.isLiveEditEnabled() ? _Editor.default.publishInfo : _Editor.default.publishInfoHidden
      }, _react.default.createElement(_reactTippy.Tooltip, {
        arrow: true,
        theme: "light",
        disabled: 'ontouchstart' in document.documentElement,
        className: _Editor.default.publishButton //title={errors.length > 0 ? 'Fix errors before publishing' : 'Publish (Ctrl+Alt+P)'}
        ,
        html: this.renderPublishButtonTooltip(errors, published)
      }, _react.default.createElement(_default2.default, {
        disabled: isReconnecting || !draft || errors.length > 0,
        onClick: this.handlePublishRequested,
        color: "primary"
      }, "Publish")), _react.default.createElement("div", {
        className: _Editor.default.publishInfoUndoButton
      }, published && _react.default.createElement(_default2.default, {
        kind: "simple",
        onClick: () => this.setState({
          showConfirmDiscard: true
        })
      }, "Discard changes"), this.state.showConfirmDiscard && _react.default.createElement(_popover.default, {
        onClickOutside: this.handleCancelDiscard,
        useOverlay: false,
        hasAnimation: true
      }, _react.default.createElement("div", null, _react.default.createElement("div", {
        className: _Editor.default.popOverText
      }, _react.default.createElement("strong", null, "Are you sure"), " you want to discard all changes since last published?"), _react.default.createElement(_buttonGrid.default, null, _react.default.createElement(_default2.default, {
        kind: "simple",
        onClick: this.handleCancelDiscard
      }, "Cancel"), _react.default.createElement(_default2.default, {
        color: "danger",
        onClick: this.handleConfirmDiscard
      }, "Discard")))), !published && _react.default.createElement(_default2.default, {
        kind: "simple",
        onClick: () => this.setState({
          showConfirmDelete: true
        })
      }, "Delete document"))));
    });
  }

  componentDidMount() {
    this.unlistenForKey = listen(window, 'keyup', event => {
      if (event.code === 'Escape' && this.state.showValidationTooltip) {
        this.setState({
          showValidationTooltip: false
        });
        return;
      }

      if (event.ctrlKey && event.code === 'KeyI' && event.altKey && !event.shiftKey) {
        this.setState(prevState => ({
          inspect: !prevState.inspect
        }));
        return;
      }

      if (event.ctrlKey && event.code === 'KeyP' && event.altKey && !event.shiftKey) {
        this.handlePublishRequested();
        return;
      }

      if (event.ctrlKey && event.code === 'KeyO' && event.altKey && !event.shiftKey) {
        var _this$props8 = this.props,
            draft = _this$props8.draft,
            published = _this$props8.published;
        var item = getProductionPreviewItem(draft || published);

        if (item && item.url) {
          navigateUrl(item.url);
        }
      }
    });

    if (_filterFieldsFn.default) {
      this.filterFieldFnSubscription = _filterFieldsFn.default.subscribe(filterField => this.setState({
        filterField
      }));
    }
  }

  componentWillUnmount() {
    this.unlistenForKey();
    this.setSavingStatus.cancel();

    if (this.filterFieldFnSubscription) {
      this.filterFieldFnSubscription.unsubscribe();
    }

    if (this.duplicate$) {
      this.duplicate$.unsubscribe();
    }
  } // @todo move publishing notification out of this component


  componentWillReceiveProps(nextProps) {
    this.setState({
      didPublish: this.props.isPublishing && !nextProps.isPublishing
    });

    if (this.props.isSaving && !nextProps.isSaving) {
      this.setState({
        showSavingStatus: true
      });
      this.setSavingStatus();
    }
  }

  isLiveEditEnabled() {
    var selectedSchemaType = _schema.default.get(this.props.type.name);

    return selectedSchemaType.liveEdit === true;
  }

  getTitle(value) {
    var _this$props9 = this.props,
        paneTitle = _this$props9.title,
        type = _this$props9.type;

    if (paneTitle) {
      return _react.default.createElement("span", null, paneTitle);
    }

    if (!value) {
      return "Creating new ".concat(type.title || type.name);
    }

    return _react.default.createElement(_preview.PreviewFields, {
      document: value,
      type: type,
      fields: ['title']
    }, (_ref) => {
      var title = _ref.title;
      return title ? _react.default.createElement("span", null, title) : _react.default.createElement("em", null, "Untitled");
    });
  }

  render() {
    var _this$props10 = this.props,
        draft = _this$props10.draft,
        markers = _this$props10.markers,
        published = _this$props10.published,
        type = _this$props10.type,
        isLoading = _this$props10.isLoading,
        isReconnecting = _this$props10.isReconnecting,
        patchChannel = _this$props10.patchChannel,
        transactionResult = _this$props10.transactionResult,
        onClearTransactionResult = _this$props10.onClearTransactionResult;
    var _this$state2 = this.state,
        inspect = _this$state2.inspect,
        focusPath = _this$state2.focusPath,
        showConfirmDelete = _this$state2.showConfirmDelete,
        showConfirmUnpublish = _this$state2.showConfirmUnpublish,
        didPublish = _this$state2.didPublish,
        filterField = _this$state2.filterField;
    var value = draft || published;

    if (isLoading) {
      return _react.default.createElement("div", {
        className: _Editor.default.loading
      }, _react.default.createElement(_spinner.default, {
        center: true,
        message: "Loading ".concat(type.title, "\u2026"),
        delay: 600
      }));
    }

    var hasTypeMismatch = value && value._type && value._type !== type.name;

    if (hasTypeMismatch) {
      return _react.default.createElement("div", {
        className: _Editor.default.typeMisMatchMessage
      }, "This document is of type ", _react.default.createElement("code", null, value._type), " and cannot be edited as", ' ', _react.default.createElement("code", null, type.name), _react.default.createElement("div", null, _react.default.createElement(_default2.default, {
        onClick: this.handleEditAsActualType
      }, "Edit as ", value._type, " instead")));
    }

    var enabledActions = (0, _documentActionUtils.resolveEnabledActions)(type);
    return _react.default.createElement(_default3.default, {
      styles: this.props.paneStyles,
      index: this.props.index,
      title: this.getTitle(value),
      onAction: this.handleMenuAction,
      menuItems: getMenuItems(enabledActions, draft, published, this.isLiveEditEnabled()),
      renderActions: this.renderActions,
      onMenuToggle: this.handleMenuToggle,
      isSelected: true // last pane is always selected for now
      ,
      staticContent: this.renderPublishInfo(),
      contentMaxWidth: 672
    }, _react.default.createElement("div", {
      className: _Editor.default.root
    }, _react.default.createElement("div", {
      className: _Editor.default.top
    }, !this.isLiveEditEnabled() && _react.default.createElement(_react.default.Fragment, null, published ? _react.default.createElement("strong", null, "Published ", _react.default.createElement(_TimeAgo.default, {
      time: published._updatedAt
    })) : _react.default.createElement("strong", null, "Not published")), value && _react.default.createElement(_react.default.Fragment, null, !this.isLiveEditEnabled() && _react.default.createElement("span", null, " \xB7 "), _react.default.createElement("span", null, "Edited ", _react.default.createElement(_TimeAgo.default, {
      time: value._updatedAt
    })))), _react.default.createElement("form", {
      className: _Editor.default.editor,
      onSubmit: preventDefault,
      id: "Sanity_Default_DeskTool_Editor_ScrollContainer"
    }, _react.default.createElement(_formBuilder.default, {
      schema: _schema.default,
      patchChannel: patchChannel,
      value: draft || published || {
        _type: type.name
      },
      type: type,
      filterField: filterField,
      readOnly: isReconnecting || !(0, _documentActionUtils.isActionEnabled)(type, 'update'),
      onBlur: this.handleBlur,
      onFocus: this.handleFocus,
      focusPath: focusPath,
      onChange: this.handleChange,
      markers: markers
    })), _afterEditorComponent.default.map((AfterEditorComponent, i) => _react.default.createElement(AfterEditorComponent, {
      key: i,
      documentId: published._id
    })), inspect && _react.default.createElement(_InspectView.default, {
      value: value,
      onClose: this.handleHideInspector
    }), showConfirmDelete && _react.default.createElement(_ConfirmDelete.default, {
      draft: draft,
      published: published,
      onCancel: this.handleCancelDelete,
      onConfirm: this.handleConfirmDelete
    }), showConfirmUnpublish && _react.default.createElement(_ConfirmUnpublish.default, {
      draft: draft,
      published: published,
      onCancel: this.handleCancelUnpublish,
      onConfirm: this.handleConfirmUnpublish
    }), isReconnecting && _react.default.createElement(_default4.default, {
      kind: "warning"
    }, _react.default.createElement(_warningIcon.default, null), " Connection lost. Reconnecting\u2026"), didPublish && _react.default.createElement(_default4.default, {
      kind: "success",
      timeout: 4
    }, _react.default.createElement(_circleCheckIcon.default, null), " You just published", ' ', _react.default.createElement("em", null, _react.default.createElement(_DocTitle.default, {
      document: draft || published
    }))), transactionResult && transactionResult.type === 'error' && _react.default.createElement(_default4.default, {
      kind: "danger",
      action: {
        title: 'Ok, got it'
      },
      onAction: onClearTransactionResult
    }, _react.default.createElement("div", null, transactionResult.message, _react.default.createElement("details", null, transactionResult.error.message)))));
  }

}, _defineProperty(_class, "propTypes", {
  title: _propTypes.default.string,
  paneIndex: _propTypes.default.number.isRequired,
  index: _propTypes.default.number,
  paneStyles: _propTypes.default.object,
  patchChannel: _propTypes.default.object,
  draft: _propTypes.default.object,
  published: _propTypes.default.object,
  type: _propTypes.default.object.isRequired,
  markers: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.array
  })),
  router: _propTypes.default.shape({
    state: _propTypes.default.object
  }).isRequired,
  onDelete: _propTypes.default.func,
  onCreate: _propTypes.default.func,
  onChange: _propTypes.default.func,
  onDiscardDraft: _propTypes.default.func,
  onPublish: _propTypes.default.func,
  onUnpublish: _propTypes.default.func,
  transactionResult: _propTypes.default.shape({
    type: _propTypes.default.string
  }),
  onClearTransactionResult: _propTypes.default.func,
  validationPending: _propTypes.default.bool,
  isCreatingDraft: _propTypes.default.bool,
  isUnpublishing: _propTypes.default.bool,
  isPublishing: _propTypes.default.bool,
  isReconnecting: _propTypes.default.bool,
  isLoading: _propTypes.default.bool,
  isSaving: _propTypes.default.bool,
  deletedSnapshot: _propTypes.default.object
}), _defineProperty(_class, "defaultProps", {
  title: null,
  markers: [],
  isLoading: false,
  isSaving: false,
  isUnpublishing: false,
  isPublishing: false,
  isReconnecting: false,
  isCreatingDraft: false,
  deletedSnapshot: null,
  transactionResult: null,

  onDelete() {},

  onCreate() {},

  onChange() {},

  onClearTransactionResult() {}

}), _temp));

exports.default = _default;
