// @sanity/desk-tool/lib/pane/PaneItemWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PaneItemWrapper;

var _PaneItem = _interopRequireDefault(require("./styles/PaneItem.css"));

var _ListView = _interopRequireDefault(require("./styles/ListView.css"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _router = require("part:@sanity/base/router");

var _grid = require("part:@sanity/components/lists/grid");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PaneItemWrapper(props) {
  var useGrid = props.useGrid,
      linkState = props.linkState,
      layout = props.layout,
      isSelected = props.isSelected;

  var link = _react.default.createElement(_router.StateLink, {
    state: linkState,
    className: isSelected ? _PaneItem.default.linkIsSelected : _PaneItem.default.link
  }, props.children);

  return useGrid ? _react.default.createElement(_grid.Item, {
    className: _ListView.default["".concat(layout, "ListItem")]
  }, link) : _react.default.createElement("div", {
    className: isSelected ? _PaneItem.default.selected : _PaneItem.default.item
  }, link);
}

PaneItemWrapper.propTypes = {
  layout: _propTypes.default.string,
  useGrid: _propTypes.default.bool,
  isSelected: _propTypes.default.bool,
  linkState: _propTypes.default.object,
  children: _propTypes.default.node
};
