// @sanity/desk-tool/lib/pane/Pane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _DocumentsListPane = _interopRequireDefault(require("./DocumentsListPane"));

var _UserComponentPane = _interopRequireDefault(require("./UserComponentPane"));

var _UnknownPaneType = _interopRequireDefault(require("./UnknownPaneType"));

var _EditorPane = _interopRequireDefault(require("./EditorPane"));

var _ListPane = _interopRequireDefault(require("./ListPane"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var paneMap = {
  list: _ListPane.default,
  documentList: _DocumentsListPane.default,
  document: _EditorPane.default,
  component: _UserComponentPane.default // eslint-disable-next-line react/prefer-stateless-function

};

class Pane extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handlePaneCollapse", () => this.props.onCollapse(this.props.index));

    _defineProperty(this, "handlePaneExpand", () => this.props.onExpand(this.props.index));
  }

  render() {
    var type = this.props.type;
    var ActualPane = paneMap[type] || _UnknownPaneType.default;
    return _react.default.createElement(ActualPane, _extends({}, this.props, {
      onExpand: this.handlePaneExpand,
      onCollapse: this.handlePaneCollapse
    }));
  }

}

exports.default = Pane;

_defineProperty(Pane, "propTypes", {
  index: _propTypes.default.number,
  title: _propTypes.default.string,
  type: _propTypes.default.string.isRequired,
  onCollapse: _propTypes.default.func,
  onExpand: _propTypes.default.func
});

_defineProperty(Pane, "defaultProps", {
  title: '',
  index: 0,
  onCollapse: _noop2.default,
  onExpand: _noop2.default
});
