// bs-rebel/lib/js/src/native/MutableQueue_native.js
'use strict';

var Belt_MutableQueue = require("bs-platform/lib/js/belt_MutableQueue.js");

function addMany(q1, q2) {
  return Belt_MutableQueue.transfer(Belt_MutableQueue.copy(q1), q2);
}

var fromArray = Belt_MutableQueue.fromArray;

var toArray = Belt_MutableQueue.toArray;

var make = Belt_MutableQueue.make;

var clear = Belt_MutableQueue.clear;

var add = Belt_MutableQueue.add;

var peek = Belt_MutableQueue.peek;

var pop = Belt_MutableQueue.pop;

var copy = Belt_MutableQueue.copy;

var size = Belt_MutableQueue.size;

var mapU = Belt_MutableQueue.mapU;

var map = Belt_MutableQueue.map;

var forEachU = Belt_MutableQueue.forEachU;

var forEach = Belt_MutableQueue.forEach;

var isEmpty = Belt_MutableQueue.isEmpty;

var reduceU = Belt_MutableQueue.reduceU;

var reduce = Belt_MutableQueue.reduce;

var transfer = Belt_MutableQueue.transfer;

exports.fromArray = fromArray;
exports.toArray = toArray;
exports.make = make;
exports.clear = clear;
exports.add = add;
exports.peek = peek;
exports.pop = pop;
exports.copy = copy;
exports.size = size;
exports.mapU = mapU;
exports.map = map;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.isEmpty = isEmpty;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.addMany = addMany;
exports.transfer = transfer;
/* No side effect */
