// wonka/lib/js/src/web/wonkaJs.js
'use strict';

var Wonka_operator_delay = require("./wonka_operator_delay.js");
var Wonka_sink_toPromise = require("./wonka_sink_toPromise.js");
var Wonka_operator_sample = require("./wonka_operator_sample.js");
var Wonka_operator_debounce = require("./wonka_operator_debounce.js");
var Wonka_operator_interval = require("./wonka_operator_interval.js");
var Wonka_operator_throttle = require("./wonka_operator_throttle.js");
var Wonka_source_fromPromise = require("./wonka_source_fromPromise.js");
var Wonka_source_fromDomEvent = require("./wonka_source_fromDomEvent.js");
var Wonka_source_fromListener = require("./wonka_source_fromListener.js");

var debounce = Wonka_operator_debounce.debounce;

var delay = Wonka_operator_delay.delay;

var interval = Wonka_operator_interval.interval;

var sample = Wonka_operator_sample.sample;

var throttle = Wonka_operator_throttle.throttle;

var toPromise = Wonka_sink_toPromise.toPromise;

var fromDomEvent = Wonka_source_fromDomEvent.fromDomEvent;

var fromListener = Wonka_source_fromListener.fromListener;

var fromPromise = Wonka_source_fromPromise.fromPromise;

exports.debounce = debounce;
exports.delay = delay;
exports.interval = interval;
exports.sample = sample;
exports.throttle = throttle;
exports.toPromise = toPromise;
exports.fromDomEvent = fromDomEvent;
exports.fromListener = fromListener;
exports.fromPromise = fromPromise;
/* No side effect */
