// wonka/src/web/wonkaJs.js


import * as Wonka_operator_delay from "./wonka_operator_delay.js";
import * as Wonka_sink_toPromise from "./wonka_sink_toPromise.js";
import * as Wonka_operator_sample from "./wonka_operator_sample.js";
import * as Wonka_operator_debounce from "./wonka_operator_debounce.js";
import * as Wonka_operator_interval from "./wonka_operator_interval.js";
import * as Wonka_operator_throttle from "./wonka_operator_throttle.js";
import * as Wonka_source_fromPromise from "./wonka_source_fromPromise.js";
import * as Wonka_source_fromDomEvent from "./wonka_source_fromDomEvent.js";
import * as Wonka_source_fromListener from "./wonka_source_fromListener.js";

var debounce = Wonka_operator_debounce.debounce;

var delay = Wonka_operator_delay.delay;

var interval = Wonka_operator_interval.interval;

var sample = Wonka_operator_sample.sample;

var throttle = Wonka_operator_throttle.throttle;

var toPromise = Wonka_sink_toPromise.toPromise;

var fromDomEvent = Wonka_source_fromDomEvent.fromDomEvent;

var fromListener = Wonka_source_fromListener.fromListener;

var fromPromise = Wonka_source_fromPromise.fromPromise;

export {
  debounce ,
  delay ,
  interval ,
  sample ,
  throttle ,
  toPromise ,
  fromDomEvent ,
  fromListener ,
  fromPromise ,
  
}
/* No side effect */
