// @sanity/base/lib/datastores/utils/Location.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.configure = configure;
exports.default = void 0;

var _url = _interopRequireDefault(require("url"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function configure() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
    qsImpl: require('querystring')
  },
      qsImpl = _ref.qsImpl;

  class Location {
    get search() {
      var stringified = qsImpl.stringify(this.query || {});
      return stringified.length > 1 ? "?".concat(stringified) : null;
    }

    set search(newSearch) {
      this.query = qsImpl.parse(newSearch.replace(/^\?/, ''));
    }

    get host() {
      return this.port ? "".concat(this.hostname, ":").concat(this.port) : this.hostname;
    }

    get path() {
      return this.pathname + (this.search || '');
    }

    set path(newPath) {
      var parsed = _url.default.parse(newPath, false, false);

      Object.assign(this, {
        pathname: parsed.pathname,
        query: qsImpl.parse(parsed.search.substring(1))
      });
    }

    set host(newVal) {
      var _newVal$split = newVal.split(':'),
          _newVal$split2 = _slicedToArray(_newVal$split, 2),
          hostname = _newVal$split2[0],
          port = _newVal$split2[1];

      Object.assign(this, {
        hostname: hostname,
        port: port
      });
    }

    get href() {
      return _url.default.format(this);
    }

    set href(newHref) {
      var parsed = _url.default.parse(newHref, false, false);

      Object.assign(this, {
        protocol: parsed.protocol,
        slashes: parsed.slashes,
        hostname: parsed.hostname,
        pathname: parsed.pathname,
        port: parsed.port,
        auth: parsed.auth,
        query: qsImpl.parse((parsed.search || '').substring(1)),
        hash: parsed.hash
      });
    }

    clone() {
      Object.assign(new Location(), this);
    }

    extend(properties) {
      return Object.assign(new Location(), this, properties);
    }

  }

  return {
    parse(urlToParse) {
      return Object.assign(new Location(), {
        href: urlToParse
      });
    },

    stringify(u) {
      return _url.default.format({
        protocol: u.protocol,
        slashes: u.slashes,
        hostname: u.hostname,
        pathname: u.pathname,
        port: u.port,
        search: u.search,
        hash: u.hash
      });
    }

  };
}

var _default = configure();

exports.default = _default;
