// @sanity/components/lib/previews/BlockImagePreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _blockImageStyle = _interopRequireDefault(require("part:@sanity/components/previews/block-image-style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var fieldProp = _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.func]);

class BlockImagePreview extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        description = _this$props.description,
        mediaDimensions = _this$props.mediaDimensions,
        media = _this$props.media,
        children = _this$props.children;
    return _react.default.createElement("div", {
      className: _blockImageStyle.default.root
    }, _react.default.createElement("h2", {
      className: _blockImageStyle.default.title
    }, title), _react.default.createElement("div", {
      className: _blockImageStyle.default.preview
    }, media && _react.default.createElement("div", {
      className: "".concat(_blockImageStyle.default.media)
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'blockImage'
    }), typeof media === 'string' && _react.default.createElement("div", {
      className: _blockImageStyle.default.mediaString
    }, media), _react.default.isValidElement(media) && media), _react.default.createElement("div", {
      className: _blockImageStyle.default.heading
    }, subtitle && _react.default.createElement("h3", {
      className: _blockImageStyle.default.subtitle
    }, subtitle), description && _react.default.createElement("p", {
      className: _blockImageStyle.default.description
    }, description), status && _react.default.createElement("div", {
      className: _blockImageStyle.default.status
    }, typeof status === 'function' && status({
      layout: 'default'
    }) || status))), children && _react.default.createElement("div", {
      className: _blockImageStyle.default.children
    }, children));
  }

}

exports.default = BlockImagePreview;

_defineProperty(BlockImagePreview, "propTypes", {
  title: _propTypes.default.string,
  subtitle: _propTypes.default.string,
  description: _propTypes.default.string,
  mediaDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number,
    fit: _propTypes.default.oneOf(['clip', 'crop', 'fill', 'fillmax', 'max', 'scale', 'min']),
    aspect: _propTypes.default.number
  }),
  media: fieldProp,
  children: _propTypes.default.func,
  type: _propTypes.default.shape({
    title: _propTypes.default.string,
    name: _propTypes.default.string
  })
});

_defineProperty(BlockImagePreview, "defaultProps", {
  mediaDimensions: {
    width: 600,
    height: 600,
    fit: 'fillmax'
  },
  type: {
    title: undefined,
    name: undefined
  }
});
