// @sanity/components/lib/previews/DefaultPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _defaultStyle = _interopRequireDefault(require("part:@sanity/components/previews/default-style"));

var _Styleable = _interopRequireDefault(require("../utilities/Styleable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var fieldProp = _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.node, _propTypes.default.func]);

class DefaultPreview extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        title = _this$props.title,
        subtitle = _this$props.subtitle,
        media = _this$props.media,
        children = _this$props.children,
        status = _this$props.status,
        isPlaceholder = _this$props.isPlaceholder,
        progress = _this$props.progress,
        mediaDimensions = _this$props.mediaDimensions,
        styles = _this$props.styles;

    if (isPlaceholder) {
      return _react.default.createElement("div", {
        className: styles.placeholder
      }, media !== false && _react.default.createElement("div", {
        className: styles.media
      }), _react.default.createElement("div", {
        className: styles.heading
      }, _react.default.createElement("h2", {
        className: styles.title
      }, "Loading\u2026"), _react.default.createElement("h3", {
        className: styles.subtitle
      }, "Loading\u2026")));
    }

    return _react.default.createElement("div", {
      className: "\n          ".concat(styles.root, "\n          ").concat(subtitle ? styles.hasSubtitle : '', "\n        ")
    }, media !== false && _react.default.createElement("div", {
      className: styles.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && _react.default.createElement("div", {
      className: styles.mediaString
    }, media), _react.default.isValidElement(media) && media), _react.default.createElement("div", {
      className: styles.heading
    }, _react.default.createElement("h2", {
      className: styles.title
    }, typeof title !== 'function' && title, typeof title === 'function' && title({
      layout: 'default'
    })), subtitle && _react.default.createElement("h3", {
      className: styles.subtitle
    }, typeof subtitle === 'function' && subtitle({
      layout: 'default'
    }) || subtitle)), status && _react.default.createElement("div", {
      className: styles.status
    }, typeof status === 'function' && status({
      layout: 'default'
    }) || status), children && _react.default.createElement("div", {
      className: styles.children
    }, children), typeof progress === 'number' && progress > -1 && _react.default.createElement("div", {
      className: styles.progress
    }, _react.default.createElement("div", {
      className: styles.progressBar,
      style: {
        width: "".concat(progress, "%")
      }
    })));
  }

}

_defineProperty(DefaultPreview, "propTypes", {
  title: fieldProp,
  subtitle: fieldProp,
  mediaDimensions: _propTypes.default.shape({
    width: _propTypes.default.number,
    height: _propTypes.default.number,
    fit: _propTypes.default.oneOf(['clip', 'crop', 'fill', 'fillmax', 'max', 'scale', 'min']),
    aspect: _propTypes.default.number
  }),
  status: fieldProp,
  media: fieldProp,
  isPlaceholder: _propTypes.default.bool,
  children: _propTypes.default.node,
  styles: _propTypes.default.object,
  progress: _propTypes.default.number
});

_defineProperty(DefaultPreview, "defaultProps", {
  title: 'Untitled',
  subtitle: undefined,
  progress: undefined,
  media: undefined,
  mediaDimensions: {
    width: 80,
    height: 80,
    aspect: 1,
    fit: 'crop'
  }
});

var _default = (0, _Styleable.default)(DefaultPreview, _defaultStyle.default);

exports.default = _default;
