// @sanity/form-builder/lib/inputs/ImageInput/ImageInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _partition4 = _interopRequireDefault(require("lodash/partition"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _button = _interopRequireDefault(require("part:@sanity/components/fileinput/button"));

var _circle = _interopRequireDefault(require("part:@sanity/components/progress/circle"));

var _editIcon = _interopRequireDefault(require("part:@sanity/base/edit-icon"));

var _uploadIcon = _interopRequireDefault(require("part:@sanity/base/upload-icon"));

var _visibilityIcon = _interopRequireDefault(require("part:@sanity/base/visibility-icon"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

var _ImageInput = _interopRequireDefault(require("./styles/ImageInput.css"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _observable = require("../../typedefs/observable");

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _WithMaterializedReference = _interopRequireDefault(require("../../utils/WithMaterializedReference"));

var _ImageToolInput = _interopRequireDefault(require("../ImageToolInput"));

var _HotspotImage = _interopRequireDefault(require("@sanity/imagetool/HotspotImage"));

var _SelectAsset = _interopRequireDefault(require("./SelectAsset"));

var _FormBuilderInput = require("../../FormBuilderInput");

var _UploadPlaceholder = _interopRequireDefault(require("../common/UploadPlaceholder"));

var _UploadTargetFieldset = _interopRequireDefault(require("../../utils/UploadTargetFieldset"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _imagetool = _interopRequireDefault(require("@sanity/imagetool"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var HIDDEN_FIELDS = ['asset', 'hotspot', 'crop'];

var getDevicePixelRatio = () => {
  if (typeof window === 'undefined' || !window.devicePixelRatio) {
    return 1;
  }

  return Math.max(1, window.devicePixelRatio);
};

class ImageInput extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "uploadSubscription", void 0);

    _defineProperty(this, "state", {
      isUploading: false,
      uploadError: null,
      isAdvancedEditOpen: false,
      isSelectAssetOpen: false,
      hasFocus: false
    });

    _defineProperty(this, "handleRemoveButtonClick", event => {
      this.props.onChange(_PatchEvent.default.from((0, _PatchEvent.unset)(['asset'])));
    });

    _defineProperty(this, "handleCancelUpload", () => {
      this.cancelUpload();
    });

    _defineProperty(this, "handleSelectFile", files => {
      this.uploadFirstAccepted(files);
    });

    _defineProperty(this, "getConstrainedImageSrc", assetDocument => {
      var materializedSize = _imagetool.default.maxWidth || 1000;
      var maxSize = materializedSize * getDevicePixelRatio();
      var constrainedSrc = "".concat(assetDocument.url, "?w=").concat(maxSize, "&h=").concat(maxSize, "&fit=max");
      return constrainedSrc;
    });

    _defineProperty(this, "renderMaterializedAsset", assetDocument => {
      var _this$props$value = this.props.value,
          value = _this$props$value === void 0 ? {} : _this$props$value;
      var constrainedSrc = this.getConstrainedImageSrc(assetDocument);
      var srcAspectRatio = (0, _get2.default)(assetDocument, 'metadata.dimensions.aspectRatio');
      return typeof srcAspectRatio === 'undefined' ? null : _react.default.createElement(_HotspotImage.default, {
        aspectRatio: "auto",
        src: constrainedSrc,
        srcAspectRatio: srcAspectRatio,
        hotspot: value.hotspot,
        crop: value.crop
      });
    });

    _defineProperty(this, "handleFieldChange", (event, field) => {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          type = _this$props.type;
      onChange(event.prefixAll(field.name).prepend((0, _PatchEvent.setIfMissing)({
        _type: type.name
      })));
    });

    _defineProperty(this, "handleStartAdvancedEdit", () => {
      this.setState({
        isAdvancedEditOpen: true
      });
    });

    _defineProperty(this, "handleStopAdvancedEdit", () => {
      this.setState({
        isAdvancedEditOpen: false
      });
    });

    _defineProperty(this, "handleOpenSelectAsset", () => {
      this.setState({
        isSelectAssetOpen: true
      });
    });

    _defineProperty(this, "handleCloseSelectAsset", () => {
      this.setState({
        isSelectAssetOpen: false
      });
    });

    _defineProperty(this, "handleSelectAsset", asset => {
      var _this$props2 = this.props,
          onChange = _this$props2.onChange,
          type = _this$props2.type;
      onChange(_PatchEvent.default.from([(0, _PatchEvent.setIfMissing)({
        _type: type.name
      }), (0, _PatchEvent.unset)(['hotspot']), (0, _PatchEvent.unset)(['crop']), (0, _PatchEvent.set)({
        _type: 'reference',
        _ref: asset._id
      }, ['asset'])]));
      this.setState({
        isSelectAssetOpen: false
      });
    });

    _defineProperty(this, "handleFocus", event => {
      this.setState({
        hasFocus: true
      });
      this.props.onFocus(event);
    });

    _defineProperty(this, "handleBlur", event => {
      this.props.onBlur(event);
      this.setState({
        hasFocus: false
      });
    });

    _defineProperty(this, "setFocusArea", el => {
      this._focusArea = el;
    });

    _defineProperty(this, "getUploadOptions", file => {
      var _this$props3 = this.props,
          type = _this$props3.type,
          resolveUploader = _this$props3.resolveUploader;
      var uploader = resolveUploader && resolveUploader(type, file);
      return uploader ? [{
        type: type,
        uploader
      }] : [];
    });

    _defineProperty(this, "handleUpload", (_ref) => {
      var file = _ref.file,
          uploader = _ref.uploader;
      this.uploadWith(uploader, file);
    });
  }

  clearUploadStatus() {
    this.props.onChange(_PatchEvent.default.from([(0, _PatchEvent.unset)(['_upload'])])); // todo: this is kind of hackish
  }

  cancelUpload() {
    if (this.uploadSubscription) {
      this.uploadSubscription.unsubscribe();
      this.clearUploadStatus();
    }
  }

  uploadFirstAccepted(fileList) {
    var _this$props4 = this.props,
        resolveUploader = _this$props4.resolveUploader,
        type = _this$props4.type;
    var match;
    Array.from(fileList).some(file => {
      var uploader = resolveUploader(type, file);

      if (uploader) {
        match = {
          file,
          uploader
        };
        return true;
      }

      return false;
    });

    if (match) {
      this.uploadWith(match.uploader, match.file);
    }
  }

  uploadWith(uploader, file) {
    var _this$props5 = this.props,
        type = _this$props5.type,
        onChange = _this$props5.onChange;
    var options = {
      metadata: (0, _get2.default)(type, 'options.metadata'),
      storeOriginalFilename: (0, _get2.default)(type, 'options.storeOriginalFilename')
    };
    this.cancelUpload();
    this.setState({
      isUploading: true
    });
    onChange(_PatchEvent.default.from([(0, _PatchEvent.setIfMissing)({
      _type: type.name
    })]));
    this.uploadSubscription = uploader.upload(file, type, options).subscribe({
      next: uploadEvent => {
        if (uploadEvent.patches) {
          onChange(_PatchEvent.default.from(uploadEvent.patches));
        }
      },
      error: err => {
        console.error(err);
        this.setState({
          uploadError: err
        });
        this.clearUploadStatus();
      },
      complete: () => {
        onChange(_PatchEvent.default.from([(0, _PatchEvent.unset)(['hotspot']), (0, _PatchEvent.unset)(['crop'])]));
        this.setState({
          isUploading: false
        });
      }
    });
  }

  renderUploadState(uploadState) {
    var isUploading = this.state.isUploading;
    var isComplete = uploadState.progress === 100 && !!(this.props.value && this.props.value.asset);
    return _react.default.createElement("div", {
      className: _ImageInput.default.progress
    }, _react.default.createElement("div", null, _react.default.createElement("div", {
      className: isComplete ? _ImageInput.default.progressBarCompleted : _ImageInput.default.progressBar
    }, _react.default.createElement(_circle.default, {
      percent: status === 'complete' ? 100 : uploadState.progress,
      text: isComplete ? 'Please wait…' : 'Uploading…',
      completed: isComplete,
      showPercent: true,
      animation: true
    })), isUploading && _react.default.createElement(_default.default, {
      kind: "simple",
      color: "danger",
      onClick: this.handleCancelUpload
    }, "Cancel")));
  }

  isImageToolEnabled() {
    return (0, _get2.default)(this.props.type, 'options.hotspot') === true;
  }

  renderAdvancedEdit(fields) {
    var _this$props6 = this.props,
        value = _this$props6.value,
        level = _this$props6.level,
        type = _this$props6.type,
        onChange = _this$props6.onChange,
        readOnly = _this$props6.readOnly,
        materialize = _this$props6.materialize;
    return _react.default.createElement(_fullscreen.default, {
      title: "Edit details",
      onClose: this.handleStopAdvancedEdit,
      isOpen: true
    }, this.isImageToolEnabled() && value && value.asset && _react.default.createElement(_WithMaterializedReference.default, {
      materialize: materialize,
      reference: value.asset
    }, imageAsset => _react.default.createElement(_ImageToolInput.default, {
      type: type,
      level: level,
      readOnly: readOnly,
      imageUrl: this.getConstrainedImageSrc(imageAsset),
      value: value,
      onChange: onChange
    })), _react.default.createElement("div", {
      className: _ImageInput.default.advancedEditFields
    }, this.renderFields(fields)), _react.default.createElement(_default.default, {
      onClick: this.handleStopAdvancedEdit
    }, "Close"));
  }

  renderFields(fields) {
    return fields.map(field => this.renderField(field));
  }

  renderField(field) {
    var _this$props7 = this.props,
        value = _this$props7.value,
        level = _this$props7.level,
        focusPath = _this$props7.focusPath,
        onFocus = _this$props7.onFocus,
        readOnly = _this$props7.readOnly,
        onBlur = _this$props7.onBlur;
    var fieldValue = value && value[field.name];
    return _react.default.createElement("div", {
      className: _ImageInput.default.field,
      key: field.name
    }, _react.default.createElement(_FormBuilderInput.FormBuilderInput, {
      value: fieldValue,
      type: field.type,
      onChange: ev => this.handleFieldChange(ev, field),
      path: [field.name],
      onFocus: onFocus,
      onBlur: onBlur,
      readOnly: readOnly || field.type.readOnly,
      focusPath: focusPath,
      level: level
    }));
  }

  focus() {
    if (this._focusArea) {
      this._focusArea.focus();
    }
  }

  render() {
    var _this$props8 = this.props,
        type = _this$props8.type,
        value = _this$props8.value,
        level = _this$props8.level,
        materialize = _this$props8.materialize,
        markers = _this$props8.markers,
        readOnly = _this$props8.readOnly;
    var _this$state = this.state,
        isAdvancedEditOpen = _this$state.isAdvancedEditOpen,
        isSelectAssetOpen = _this$state.isSelectAssetOpen,
        uploadError = _this$state.uploadError,
        hasFocus = _this$state.hasFocus;

    var _partition2 = (0, _partition4.default)(type.fields.filter(field => !HIDDEN_FIELDS.includes(field.name)), 'type.options.isHighlighted'),
        _partition3 = _slicedToArray(_partition2, 2),
        highlightedFields = _partition3[0],
        otherFields = _partition3[1];

    var hasAsset = value && value.asset;
    var showAdvancedEditButton = value && (otherFields.length > 0 || hasAsset && this.isImageToolEnabled());
    return _react.default.createElement(_UploadTargetFieldset.default, {
      markers: markers,
      legend: type.title,
      description: type.description,
      level: level,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      className: _ImageInput.default.root,
      onUpload: this.handleUpload,
      getUploadOptions: this.getUploadOptions,
      ref: this.setFocusArea
    }, uploadError && _react.default.createElement(_default2.default, {
      kind: "error",
      action: {
        title: 'OK'
      },
      onAction: () => this.setState({
        uploadError: null
      })
    }, "We're", " really sorry, but the upload could not be completed."), _react.default.createElement("div", {
      className: _ImageInput.default.content
    }, _react.default.createElement("div", {
      className: _ImageInput.default.assetWrapper
    }, value && value._upload && _react.default.createElement("div", {
      className: _ImageInput.default.uploadState
    }, this.renderUploadState(value._upload)), hasAsset ? _react.default.createElement(_WithMaterializedReference.default, {
      reference: value.asset,
      materialize: materialize
    }, this.renderMaterializedAsset) : readOnly ? _react.default.createElement("span", null, "Field is read only") : _react.default.createElement(_UploadPlaceholder.default, {
      hasFocus: hasFocus
    }))), _react.default.createElement("div", {
      className: _ImageInput.default.functions
    }, _react.default.createElement(_buttonGrid.default, null, !readOnly && _react.default.createElement(_button.default, {
      icon: _uploadIcon.default,
      inverted: true,
      onSelect: this.handleSelectFile,
      accept: ''
      /* todo build from this.props.resolveUploaders */

    }, "Upload"), !readOnly && _react.default.createElement(_default.default, {
      onClick: this.handleOpenSelectAsset,
      inverted: true
    }, "Browse"), showAdvancedEditButton && _react.default.createElement(_default.default, {
      icon: readOnly ? _visibilityIcon.default : _editIcon.default,
      inverted: true,
      title: readOnly ? 'View details' : 'Edit details',
      onClick: this.handleStartAdvancedEdit
    }, readOnly ? 'View details' : 'Edit'), hasAsset && !readOnly && _react.default.createElement(_default.default, {
      color: "danger",
      inverted: true,
      onClick: this.handleRemoveButtonClick
    }, "Remove"))), highlightedFields.length > 0 && _react.default.createElement("div", {
      className: _ImageInput.default.fieldsWrapper
    }, this.renderFields(highlightedFields)), isAdvancedEditOpen && this.renderAdvancedEdit(otherFields), isSelectAssetOpen && _react.default.createElement(_fullscreen.default, {
      title: "Select image",
      onClose: this.handleCloseSelectAsset,
      isOpen: true
    }, _react.default.createElement(_SelectAsset.default, {
      onSelect: this.handleSelectAsset
    })));
  }

}

exports.default = ImageInput;
