// @sanity/form-builder/lib/sanity/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _SanityFormBuilder.default;
  }
});
Object.defineProperty(exports, "FormBuilder", {
  enumerable: true,
  get: function get() {
    return _SanityFormBuilder.default;
  }
});
Object.defineProperty(exports, "FormBuilderContext", {
  enumerable: true,
  get: function get() {
    return _SanityFormBuilderContext.default;
  }
});
Object.defineProperty(exports, "WithFormBuilderValue", {
  enumerable: true,
  get: function get() {
    return _WithFormBuilderValue.default;
  }
});
Object.defineProperty(exports, "withDocument", {
  enumerable: true,
  get: function get() {
    return _withDocument.default;
  }
});
Object.defineProperty(exports, "withValuePath", {
  enumerable: true,
  get: function get() {
    return _withValuePath.default;
  }
});
Object.defineProperty(exports, "FormBuilderInput", {
  enumerable: true,
  get: function get() {
    return _FormBuilderInput.FormBuilderInput;
  }
});
Object.defineProperty(exports, "checkout", {
  enumerable: true,
  get: function get() {
    return _formBuilderValueStore.checkout;
  }
});
Object.defineProperty(exports, "checkoutPair", {
  enumerable: true,
  get: function get() {
    return _formBuilderValueStore.checkoutPair;
  }
});
Object.defineProperty(exports, "PatchEvent", {
  enumerable: true,
  get: function get() {
    return _PatchEvent.default;
  }
});
Object.defineProperty(exports, "HashFocusManager", {
  enumerable: true,
  get: function get() {
    return _HashFocusManager.default;
  }
});
Object.defineProperty(exports, "SimpleFocusManager", {
  enumerable: true,
  get: function get() {
    return _SimpleFocusManager.default;
  }
});
Object.defineProperty(exports, "BlockEditor", {
  enumerable: true,
  get: function get() {
    return _BlockEditor.default;
  }
});
exports.patches = void 0;

var patches = _interopRequireWildcard(require("../utils/patches"));

exports.patches = patches;

var _SanityFormBuilder = _interopRequireDefault(require("./SanityFormBuilder"));

var _SanityFormBuilderContext = _interopRequireDefault(require("./SanityFormBuilderContext"));

var _WithFormBuilderValue = _interopRequireDefault(require("./WithFormBuilderValue"));

var _withDocument = _interopRequireDefault(require("../utils/withDocument"));

var _withValuePath = _interopRequireDefault(require("../utils/withValuePath"));

var _FormBuilderInput = require("../FormBuilderInput");

var _formBuilderValueStore = require("./formBuilderValueStore");

var _PatchEvent = _interopRequireDefault(require("../PatchEvent"));

var _HashFocusManager = _interopRequireDefault(require("./focusManagers/HashFocusManager"));

var _SimpleFocusManager = _interopRequireDefault(require("./focusManagers/SimpleFocusManager"));

var _BlockEditor = _interopRequireDefault(require("../inputs/BlockEditor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }
