// @sanity/form-builder/lib/inputs/Slug/SlugInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _get2 = _interopRequireDefault(require("lodash/get"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _speakingurl = _interopRequireDefault(require("speakingurl"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _default4 = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _FormBuilderPropTypes = _interopRequireDefault(require("../../FormBuilderPropTypes"));

var _PatchEvent = require("../../PatchEvent");

var _withDocument = _interopRequireDefault(require("../../utils/withDocument"));

var _withValuePath = _interopRequireDefault(require("../../utils/withValuePath"));

var _SlugInput = _interopRequireDefault(require("./styles/SlugInput.css"));

var _class, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Fallback slugify function if not defined in field options
function defaultSlugify(value, type) {
  var maxLength = type.options && type.options.maxLength || 200;
  var slugifyOpts = {
    truncate: maxLength,
    symbols: true
  };
  return value ? (0, _speakingurl.default)(value, slugifyOpts) : '';
}

var defaultState = {
  inputText: undefined,
  loading: false
};

var _default = (0, _withValuePath.default)((0, _withDocument.default)((_temp = _class = class SlugInput extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", defaultState);

    _defineProperty(this, "setTextInput", input => {
      this._textInput = input;
    });

    _defineProperty(this, "handleChange", event => {
      this.updateCurrent(event.target.value);
    });

    _defineProperty(this, "handleFocusCurrent", event => {
      this.props.onFocus(['current']);
    });

    _defineProperty(this, "handleGenerateSlug", () => {
      var type = this.props.type;
      var source = (0, _get2.default)(type, 'options.source');

      if (!source) {
        // eslint-disable-next-line no-console
        console.error("Source is missing. Check source on type \"".concat(type.name, "\" in schema"));
        return;
      }

      var newFromSource = this.getNewFromSource();
      this.setState({
        loading: true
      });
      this.slugify(newFromSource || '').then(newSlug => this.updateCurrent(newSlug)).catch(err => {
        // eslint-disable-next-line no-console
        console.error("An error occured while slugifying \"".concat(newFromSource, "\":\n").concat(err.message, "\n").concat(err.stack));
      }).then(() => this._isMounted && this.setState({
        loading: false
      }));
    });

    _defineProperty(this, "getNewFromSource", () => {
      var _this$props = this.props,
          getValuePath = _this$props.getValuePath,
          type = _this$props.type,
          document = _this$props.document;
      var parentPath = getValuePath().slice(0, -1);
      var source = (0, _get2.default)(type, 'options.source');
      return typeof source === 'function' ? source(document, {
        parentPath
      }) : (0, _get2.default)(document, source);
    });
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  updateCurrent(current) {
    var _this$props2 = this.props,
        onChange = _this$props2.onChange,
        type = _this$props2.type;

    if (!current) {
      onChange(_PatchEvent.PatchEvent.from((0, _PatchEvent.unset)(['current'])));
      return;
    }

    onChange(_PatchEvent.PatchEvent.from((0, _PatchEvent.setIfMissing)({
      _type: type.name
    }), (0, _PatchEvent.set)(current, ['current'])));
  }

  slugify(sourceValue) {
    if (!sourceValue) {
      return Promise.resolve(sourceValue);
    }

    var type = this.props.type;
    var slugify = (0, _get2.default)(type, 'options.slugify', defaultSlugify);
    return Promise.resolve(slugify(sourceValue, type));
  }

  componentWillReceiveProps(nextProps) {
    var document = nextProps.document; // Reset state if document is changed

    var oldDocId = this.props.document._id;
    var newDocId = document._id;

    if (oldDocId !== newDocId) {
      this.setState(defaultState);
    }
  }

  focus() {
    if (this._textInput) {
      this._textInput.focus();
    }
  }

  render() {
    var _this$props3 = this.props,
        value = _this$props3.value,
        type = _this$props3.type,
        level = _this$props3.level,
        markers = _this$props3.markers;
    var _this$state = this.state,
        loading = _this$state.loading,
        inputText = _this$state.inputText;
    var hasSourceField = type.options && type.options.source;
    var formFieldProps = {
      label: type.title,
      description: type.description,
      level: level,
      markers
    };
    var validation = markers.filter(marker => marker.type === 'validation');
    var errors = validation.filter(marker => marker.level === 'error');
    var hasSource = Boolean(this.getNewFromSource());
    return _react.default.createElement(_default3.default, formFieldProps, _react.default.createElement("div", {
      className: _SlugInput.default.wrapper
    }, _react.default.createElement("div", {
      className: _SlugInput.default.input
    }, _react.default.createElement(_default4.default, {
      ref: this.setTextInput,
      customValidity: errors.length > 0 ? errors[0].item.message : '',
      disabled: loading,
      placeholder: type.placeholder,
      onChange: this.handleChange,
      onFocus: this.handleFocusCurrent,
      value: typeof inputText === 'string' ? inputText : value.current
    })), hasSourceField && _react.default.createElement(_default2.default, {
      className: _SlugInput.default.button,
      inverted: true,
      disabled: loading || !hasSource,
      loading: loading,
      onClick: this.handleGenerateSlug
    }, "Generate")));
  }

}, _defineProperty(_class, "propTypes", {
  type: _FormBuilderPropTypes.default.type.isRequired,
  level: _propTypes.default.number.isRequired,
  value: _propTypes.default.shape({
    current: _propTypes.default.string
  }),
  document: _propTypes.default.shape({
    _id: _propTypes.default.string
  }).isRequired,
  onChange: _propTypes.default.func,
  onFocus: _propTypes.default.func.isRequired,
  getValuePath: _propTypes.default.func.isRequired,
  markers: _propTypes.default.arrayOf(_propTypes.default.shape({
    type: _propTypes.default.string.isRequired
  }))
}), _defineProperty(_class, "defaultProps", {
  value: {
    current: undefined
  },

  onChange() {},

  markers: []
}), _temp)));

exports.default = _default;
